// zarm/lib/back-to-top/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _reactDom = require("react-dom");

var _classnames = _interopRequireDefault(require("classnames"));

var _scroller = _interopRequireDefault(require("../scroller"));

var _dom = require("../utils/dom");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var BackToTop = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(BackToTop, _PureComponent);

  var _super = _createSuper(BackToTop);

  function BackToTop() {
    var _this;

    (0, _classCallCheck2.default)(this, BackToTop);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.state = {
      visible: false
    };
    _this.timer = void 0;
    _this.portalContainer = void 0;
    _this.scroller = void 0;

    _this.onScroll = function (scrollTop) {
      _this.setState({
        visible: scrollTop > _this.props.visibleDistance
      });
    };

    _this.scrollToTop = function (e) {
      var _this$props = _this.props,
          speed = _this$props.speed,
          onClick = _this$props.onClick;

      var _assertThisInitialize = (0, _assertThisInitialized2.default)(_this),
          container = _assertThisInitialize.container;

      if (typeof onClick === 'function') {
        onClick(e);
      } // 速度设置为0或者无穷大时，直接到顶


      if (speed === 0 || speed === Infinity) {
        (0, _dom.scrollTo)(container, 0, 0, 0);
        return;
      }

      var x = _this.props.speed;
      (0, _dom.scrollTo)(container, 0, 0, _this.scrollTop / (x / 10 * 1000));
    };

    return _this;
  }

  (0, _createClass2.default)(BackToTop, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.parent.appendChild(this.getPortalContainer);
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      var scrollContainer = this.props.scrollContainer;

      if (prevProps.scrollContainer !== scrollContainer) {
        this.parent.appendChild(this.portalContainer);
        this.onScroll(this.scrollTop);
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      clearInterval(this.timer);
      this.parent && this.parent.removeChild(this.portalContainer);
    }
  }, {
    key: "parent",
    get: function get() {
      if (!_dom.canUseDOM || this.container === window) {
        return document.body;
      }

      return this.container;
    }
  }, {
    key: "container",
    get: function get() {
      return this.scroller ? this.scroller.scrollContainer : window;
    }
  }, {
    key: "scrollTop",
    get: function get() {
      return this.scroller ? this.scroller.scrollTop : 0;
    }
  }, {
    key: "renderPortal",
    get: function get() {
      var _this2 = this;

      var _this$props2 = this.props,
          prefixCls = _this$props2.prefixCls,
          style = _this$props2.style,
          scrollContainer = _this$props2.scrollContainer,
          children = _this$props2.children;
      var visible = this.state.visible;

      var containerStyle = _objectSpread({
        display: !visible ? 'none' : 'inline',
        position: this.container !== window ? 'absolute' : 'fixed',
        bottom: 50,
        right: 50
      }, style);

      return /*#__PURE__*/(0, _reactDom.createPortal)( /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
        className: prefixCls,
        style: containerStyle,
        onClick: this.scrollToTop
      }, children), /*#__PURE__*/_react.default.createElement(_scroller.default, {
        ref: function ref(ele) {
          _this2.scroller = ele;
        },
        container: scrollContainer,
        onScroll: this.onScroll
      })), this.getPortalContainer);
    }
  }, {
    key: "getPortalContainer",
    get: function get() {
      var _this$props3 = this.props,
          prefixCls = _this$props3.prefixCls,
          className = _this$props3.className;

      if (!this.portalContainer) {
        var container = document.createElement('div');
        container.className = (0, _classnames.default)("".concat(prefixCls, "-container"), className);
        this.portalContainer = container;
      }

      return this.portalContainer;
    }
  }, {
    key: "render",
    value: function render() {
      if (!_dom.canUseDOM) {
        return null;
      }

      return this.renderPortal;
    }
  }]);
  return BackToTop;
}(_react.PureComponent);

exports.default = BackToTop;
BackToTop.displayName = 'BackToTop';
BackToTop.defaultProps = {
  prefixCls: 'za-back-to-top',
  speed: 100,
  visibleDistance: 400,
  scrollContainer: _dom.canUseDOM ? window : undefined
};
