// rnbridge-link/.history/src/index_20210713155456.js
/*
 * @Author: your name
 * @Date: 2021-07-13 15:54:29
 * @LastEditTime: 2021-07-13 15:54:56
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: /rnBridge/src/index.js
 */
/*
 * @Author: your name
 * @Date: 2021-06-21 14:48:22
 * @LastEditTime: 2021-07-13 14:44:50
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: /mall/src/utils/bridge.js
 */
import ReactNative from "react-native";

class NativeBridgeCaller {
  caller = ReactNative.NativeModules.UNPJSBridge.call;
  formatParams(params) {
    let format;
    Object.keys(params).map((k) => {
      const value = params[k];
      if (typeof value === "object") {
        params[k] = JSON.stringify(value);
      }
    });

    try {
      format = encodeURIComponent(JSON.stringify(params));
    } catch (e) {
      format = encodeURIComponent(JSON.stringify({}));
    }
    return format;
  }
  parseResult(result) {
    let res;
    try {
      res = JSON.parse(decodeURIComponent(result));
    } catch (e) {
      res = {
        success: false,
        error: {
          message: e.toString(),
        },
      };
    }
    return res;
  }
  async call(api, params = {}) {
    const uri = `unipal://${api}?params=${this.formatParams(params)}`;
    return this.caller(uri).then(
      (res) => {
        const p = this.parseResult(res);
        if (p.success) {
          return Promise.resolve(p.data);
        }
        return Promise.reject(p.error);
      },
      (e) => {
        const res = {
          success: false,
          error: {
            message: e.toString(),
          },
        };
        return Promise.reject(res);
      }
    );
  }
}
class NativeBridge {
  caller = new NativeBridgeCaller();
}
export default new NativeBridge();
