// zarm/lib/stepper/index.native.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _index = _interopRequireDefault(require("./style/index.native"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var getValue = function getValue(props, defaultValue) {
  if ('value' in props) {
    return props.value;
  }

  if ('defaultValue' in props) {
    return props.defaultValue;
  }

  return defaultValue;
};

var stepperStyles = _reactNative.StyleSheet.create(_index.default);

var Stepper = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(Stepper, _PureComponent);

  var _super = _createSuper(Stepper);

  function Stepper(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Stepper);
    _this = _super.call(this, props);

    _this.onInputChange = function (value) {
      value = Number(value);
      var onInputChange = _this.props.onInputChange;

      _this.setState({
        value: value
      });

      if (typeof onInputChange === 'function') {
        onInputChange(value);
      }
    };

    _this.onInputBlur = function (value) {
      var _this$props = _this.props,
          min = _this$props.min,
          max = _this$props.max,
          onChange = _this$props.onChange;
      value = Number(value);

      if (value === '' || Number.isNaN(value)) {
        value = _this.state.lastValue;
      }

      if (min !== null && value < min) {
        value = min;
      }

      if (max !== null && value > max) {
        value = max;
      }

      _this.setState({
        value: value,
        lastValue: value
      });

      if (typeof onChange === 'function') {
        onChange(value);
      }
    };

    _this.onMinusClick = function () {
      var step = _this.props.step;
      var value = _this.state.value;

      if (_this.isMinusDisabled()) {
        return;
      }

      var newValue = Number(value) - step;

      _this.onInputBlur(newValue);
    };

    _this.onPlusClick = function () {
      var step = _this.props.step;
      var value = _this.state.value;

      if (_this.isPlusDisabled()) {
        return;
      }

      var newValue = Number(value) + step;

      _this.onInputBlur(newValue);
    };

    _this.isMinusDisabled = function () {
      var _this$props2 = _this.props,
          min = _this$props2.min,
          disabled = _this$props2.disabled;
      var value = _this.state.value;

      if (min === null) {
        return false;
      }

      return value <= min || disabled;
    };

    _this.isPlusDisabled = function () {
      var _this$props3 = _this.props,
          max = _this$props3.max,
          disabled = _this$props3.disabled;
      var value = _this.state.value;

      if (max === null) {
        return false;
      }

      return value >= max || disabled;
    };

    _this.renderButton = function (type) {
      var _this$props4 = _this.props,
          shape = _this$props4.shape,
          styles = _this$props4.styles;
      var isDisabled = type === 'minus' ? _this.isMinusDisabled() : _this.isPlusDisabled();
      var buttonStyle = [styles.button, styles["".concat(shape, "Button")], isDisabled && styles.disabledButton];
      var buttonTextStyle = [styles.buttonText, isDisabled && styles.disabledText];
      return /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
        activeOpacity: 0.6,
        style: buttonStyle,
        disabled: isDisabled,
        onPress: type === 'minus' ? _this.onMinusClick : _this.onPlusClick
      }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
        style: buttonTextStyle
      }, type === 'minus' ? '-' : '+'));
    };

    _this.renderInput = function () {
      var _this$props5 = _this.props,
          disabled = _this$props5.disabled,
          styles = _this$props5.styles;
      var inputStyle = [styles.input, disabled && styles.disabledText];
      return /*#__PURE__*/_react.default.createElement(_reactNative.TextInput, {
        style: inputStyle,
        keyboardType: "numeric",
        value: _this.state.value.toString(),
        onChangeText: function onChangeText(value) {
          return _this.onInputChange(value);
        },
        onEndEditing: function onEndEditing() {
          return _this.onInputBlur(_this.state.value);
        },
        editable: !disabled,
        underlineColorAndroid: "transparent",
        returnKeyType: "done"
      });
    };

    _this.state = {
      value: getValue(props, 0),
      lastValue: getValue(props, 0)
    };
    return _this;
  }

  (0, _createClass2.default)(Stepper, [{
    key: "render",
    value: function render() {
      var _this$props6 = this.props,
          styles = _this$props6.styles,
          style = _this$props6.style;
      var wrapperStyle = [styles.container, style];
      return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: wrapperStyle
      }, this.renderButton('minus'), this.renderInput(), this.renderButton('plus'));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, state) {
      if ('value' in nextProps && nextProps.value !== state.prevValue) {
        var value = getValue(nextProps, 0);
        return {
          value: value,
          lastValue: value,
          prevValue: value
        };
      }

      return null;
    }
  }]);
  return Stepper;
}(_react.PureComponent);

exports.default = Stepper;
Stepper.defaultProps = {
  disabled: false,
  step: 1,
  styles: stepperStyles
};
