// zarm/lib/collapse/Collapse.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _classnames2 = _interopRequireDefault(require("classnames"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var getActiveKey = function getActiveKey(props) {
  var multiple = props.multiple,
      activeKey = props.activeKey,
      defaultActiveKey = props.defaultActiveKey;
  var value;

  if (typeof activeKey !== 'undefined') {
    value = activeKey;
  }

  if (!('activeKey' in props) && typeof defaultActiveKey !== 'undefined') {
    value = defaultActiveKey;
  }

  if (value) {
    return multiple ? [].concat(value) : value;
  }

  return multiple ? [] : undefined;
};

var Collapse = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Collapse, _Component);

  var _super = _createSuper(Collapse);

  function Collapse() {
    var _this;

    (0, _classCallCheck2.default)(this, Collapse);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.state = {
      activeKey: getActiveKey(_this.props)
    };

    _this.onItemChange = function (onItemChange, key) {
      if (!key) {
        return;
      }

      var activeKey = _this.state.activeKey;
      var _this$props = _this.props,
          onChange = _this$props.onChange,
          multiple = _this$props.multiple;
      var isActive;
      var newActiveKey;

      if (multiple) {
        newActiveKey = [];
        activeKey = activeKey || [];

        if (activeKey.indexOf(key) > -1) {
          newActiveKey = activeKey.filter(function (i) {
            return i !== key;
          });
        } else {
          newActiveKey = activeKey.slice(0);
          newActiveKey.push(key);
        }

        isActive = newActiveKey.indexOf(key) > -1;
      } else {
        activeKey = activeKey;
        newActiveKey = activeKey === key ? undefined : key;
        isActive = activeKey === key;
      }

      if (!('activeKey' in _this.props)) {
        _this.setState({
          activeKey: newActiveKey
        });
      }

      typeof onItemChange === 'function' && onItemChange(isActive);
      typeof onChange === 'function' && onChange(newActiveKey);
    };

    _this.renderItems = function () {
      var activeKey = _this.state.activeKey;
      var _this$props2 = _this.props,
          multiple = _this$props2.multiple,
          animated = _this$props2.animated;
      return _react.Children.map(_this.props.children, function (ele) {
        var _ele$props = ele.props,
            disabled = _ele$props.disabled,
            _onChange = _ele$props.onChange;
        var key = ele.key;
        var isActive = multiple ? (activeKey || []).indexOf(key) > -1 : activeKey === key;
        return /*#__PURE__*/(0, _react.cloneElement)(ele, {
          animated: animated,
          isActive: isActive,
          onChange: function onChange() {
            return !disabled && _this.onItemChange(_onChange, key);
          }
        });
      });
    };

    return _this;
  }

  (0, _createClass2.default)(Collapse, [{
    key: "render",
    value: function render() {
      var _this$props3 = this.props,
          prefixCls = _this$props3.prefixCls,
          className = _this$props3.className,
          onChange = _this$props3.onChange,
          animated = _this$props3.animated,
          activeKey = _this$props3.activeKey,
          defaultActiveKey = _this$props3.defaultActiveKey,
          rest = (0, _objectWithoutProperties2.default)(_this$props3, ["prefixCls", "className", "onChange", "animated", "activeKey", "defaultActiveKey"]);
      var cls = (0, _classnames2.default)(prefixCls, className, (0, _defineProperty2.default)({}, "".concat(prefixCls, "--animated"), animated));
      return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
        className: cls
      }, rest), this.renderItems());
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, state) {
      var newState = {};

      if ('activeKey' in nextProps && nextProps.activeKey !== state.activeKey) {
        newState.activeKey = getActiveKey(nextProps);
      }

      if ('activeKey' in newState) {
        return newState;
      }

      return null;
    }
  }]);
  return Collapse;
}(_react.Component);

exports.default = Collapse;
Collapse.defaultProps = {
  prefixCls: 'za-collapse',
  multiple: false,
  animated: false,
  onChange: function onChange() {}
};
Collapse.Item = void 0;
