// zarm/lib/collapse/Collapse.native.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _index = _interopRequireDefault(require("./style/index.native"));

var _validate = require("../utils/validate");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var collapseStyles = _reactNative.StyleSheet.create(_index.default);

var isPropEqual = function isPropEqual(cur, next) {
  if ((0, _validate.isArray)(next) && (0, _validate.isArray)(cur)) {
    return next.length === cur.length && next.every(function (key, i) {
      return key === cur[i];
    });
  }

  return cur === next;
};

var getActiveKey = function getActiveKey(props) {
  var activeKey = props.activeKey,
      defaultActiveKey = props.defaultActiveKey,
      multiple = props.multiple;
  var defaultKey = activeKey || activeKey === 0 ? activeKey : defaultActiveKey;

  if (defaultKey || defaultKey === 0) {
    if ((0, _validate.isArray)(defaultKey)) {
      return !multiple ? [String(defaultKey[0])] : defaultKey.map(function (key) {
        return String(key);
      });
    }

    return [String(defaultKey)];
  }

  return [];
};

var Collapse = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Collapse, _Component);

  var _super = _createSuper(Collapse);

  function Collapse(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Collapse);
    _this = _super.call(this, props);

    _this.onChange = function (onItemChange, key) {
      var _this$props = _this.props,
          multiple = _this$props.multiple,
          onChange = _this$props.onChange;
      var activeKey = _this.state.activeKey;
      var hasKey = activeKey.indexOf(key) > -1;
      var newactiveKey = [];

      if (multiple) {
        if (hasKey) {
          newactiveKey = activeKey.filter(function (i) {
            return i !== key;
          });
        } else {
          newactiveKey = activeKey.slice(0);
          newactiveKey.push(key);
        }
      } else {
        newactiveKey = hasKey ? [] : [key];
      }

      if (typeof onItemChange === 'function') {
        var isActive = newactiveKey.indexOf(key) > -1;
        onItemChange(isActive);
      }

      _this.setState({
        activeKey: newactiveKey
      });

      onChange(key);
    };

    _this.state = {
      activeKey: getActiveKey(props)
    };
    return _this;
  }

  (0, _createClass2.default)(Collapse, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps, nextState) {
      return !isPropEqual(this.props, nextProps) || !isPropEqual(this.state, nextState);
    }
  }, {
    key: "renderItems",
    value: function renderItems() {
      var _this2 = this;

      var _this$props2 = this.props,
          animated = _this$props2.animated,
          children = _this$props2.children;
      var activeKey = this.state.activeKey;
      return _react.Children.map(children, function (ele) {
        var _ele$props = ele.props,
            disabled = _ele$props.disabled,
            onChange = _ele$props.onChange;
        var key = ele.key;
        var isActive = activeKey.indexOf(key) > -1;
        return /*#__PURE__*/(0, _react.cloneElement)(ele, {
          key: key,
          isActive: isActive,
          animated: animated,
          onChange: disabled ? function () {} : function () {
            return _this2.onChange(onChange, key);
          }
        });
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props3 = this.props,
          style = _this$props3.style,
          styles = _this$props3.styles;
      var wrapperStyle = [styles.container, style];
      return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: wrapperStyle
      }, this.renderItems());
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, state) {
      var newState = {};

      if ('activeKey' in nextProps && nextProps.activeKey !== state.prevActiveKey) {
        newState.activeKey = getActiveKey(nextProps);
        newState.prevActiveKey = nextProps.activeKey;
      }

      if ('animated' in nextProps) {
        newState.animated = nextProps.animated;
      }

      if ('multiple' in nextProps) {
        newState.multiple = nextProps.multiple;
      }

      return newState.activeKey || newState.animated || newState.multiple ? newState : null;
    }
  }]);
  return Collapse;
}(_react.Component);

exports.default = Collapse;
Collapse.defaultProps = {
  multiple: false,
  animated: false,
  styles: collapseStyles,
  onChange: function onChange() {}
};
Collapse.Item = void 0;
