// zarm/lib/collapse/CollapseItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _classnames2 = _interopRequireDefault(require("classnames"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var CollapseItem = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(CollapseItem, _PureComponent);

  var _super = _createSuper(CollapseItem);

  function CollapseItem() {
    var _this;

    (0, _classCallCheck2.default)(this, CollapseItem);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.content = void 0;
    _this.state = {
      active: _this.props.isActive
    };

    _this.onClickItem = function () {
      var _this$props = _this.props,
          onChange = _this$props.onChange,
          disabled = _this$props.disabled;
      var active = _this.state.active;
      if (disabled) return;
      typeof onChange === 'function' && onChange(active);
    };

    _this.setStyle = function () {
      if (!_this.content) return;
      var active = _this.state.active;
      _this.content.style.height = active ? "".concat(_this.getContentHeight(_this.content), "px") : '0px';
    };

    _this.getContentHeight = function (content) {
      var children = (0, _toConsumableArray2.default)(content.children);
      return children.reduce(function (res, next) {
        res += next.offsetHeight;
        return res;
      }, 0);
    };

    return _this;
  }

  (0, _createClass2.default)(CollapseItem, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.setStyle();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      this.setStyle();
    }
  }, {
    key: "render",
    value: function render() {
      var _classnames,
          _this2 = this;

      var _this$props2 = this.props,
          prefixCls = _this$props2.prefixCls,
          title = _this$props2.title,
          className = _this$props2.className,
          disabled = _this$props2.disabled,
          animated = _this$props2.animated,
          isActive = _this$props2.isActive,
          children = _this$props2.children,
          onChange = _this$props2.onChange,
          rest = (0, _objectWithoutProperties2.default)(_this$props2, ["prefixCls", "title", "className", "disabled", "animated", "isActive", "children", "onChange"]);
      var active = this.state.active;
      var cls = (0, _classnames2.default)(prefixCls, className, (_classnames = {}, (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--active"), active), (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--disabled"), disabled), _classnames));
      return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
        className: cls
      }, rest), /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__header"),
        onClick: this.onClickItem
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__title")
      }, title), /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__arrow")
      })), /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__content"),
        ref: function ref(content) {
          _this2.content = content;
        }
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__content__inner")
      }, children)));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, state) {
      if ('isActive' in nextProps && nextProps.isActive !== state.prevActive) {
        return {
          active: nextProps.isActive,
          prevActive: nextProps.isActive
        };
      }

      return null;
    }
  }]);
  return CollapseItem;
}(_react.PureComponent);

exports.default = CollapseItem;
CollapseItem.defaultProps = {
  prefixCls: 'za-collapse-item',
  animated: false,
  disabled: false
};
