// zarm/lib/panel/index.native.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _index = _interopRequireDefault(require("./style/index.native"));

var _renderWithText = require("../utils/renderWithText.native");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var panelStyles = _reactNative.StyleSheet.create(_index.default);

var Panel = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(Panel, _PureComponent);

  var _super = _createSuper(Panel);

  function Panel() {
    (0, _classCallCheck2.default)(this, Panel);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Panel, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          title = _this$props.title,
          more = _this$props.more,
          style = _this$props.style,
          styles = _this$props.styles,
          children = _this$props.children;
      var wrapperStyle = [styles.container, style];
      var headerStyle = [styles.panelHeader];
      var bodyStyle = [styles.panelBody];
      return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: wrapperStyle
      }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: headerStyle
      }, /*#__PURE__*/_react.default.createElement(_renderWithText.RenderWithText, {
        component: title,
        viewStyle: panelStyles.panelHeaderTitle,
        textStyle: panelStyles.panelHeaderTitleText
      }), /*#__PURE__*/_react.default.createElement(_renderWithText.RenderWithText, {
        component: more,
        viewStyle: panelStyles.panelHeaderMore,
        textStyle: panelStyles.panelHeaderMoreText
      })), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: bodyStyle
      }, children));
    }
  }]);
  return Panel;
}(_react.PureComponent);

exports.default = Panel;
Panel.defaultProps = {
  styles: panelStyles
};
