// zarm/lib/modal/Modal.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _popup = _interopRequireDefault(require("../popup"));

var _ModalHeader = _interopRequireDefault(require("./ModalHeader"));

var _ModalBody = _interopRequireDefault(require("./ModalBody"));

var _ModalFooter = _interopRequireDefault(require("./ModalFooter"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var Modal = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Modal, _Component);

  var _super = _createSuper(Modal);

  function Modal() {
    (0, _classCallCheck2.default)(this, Modal);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Modal, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          prefixCls = _this$props.prefixCls,
          className = _this$props.className,
          shape = _this$props.shape,
          children = _this$props.children,
          maskClosable = _this$props.maskClosable,
          title = _this$props.title,
          closable = _this$props.closable,
          footer = _this$props.footer,
          onCancel = _this$props.onCancel,
          others = (0, _objectWithoutProperties2.default)(_this$props, ["prefixCls", "className", "shape", "children", "maskClosable", "title", "closable", "footer", "onCancel"]);
      var cls = {
        modal: (0, _classnames2.default)(prefixCls, className, (0, _defineProperty2.default)({}, "".concat(prefixCls, "--").concat(shape), !!shape)),
        dialog: (0, _classnames2.default)("".concat(prefixCls, "__dialog"))
      };
      var showHeader = title || closable;

      var noop = function noop() {};

      return /*#__PURE__*/_react.default.createElement(_popup.default, (0, _extends2.default)({
        className: cls.modal,
        direction: "center",
        onMaskClick: maskClosable ? onCancel : noop
      }, others), /*#__PURE__*/_react.default.createElement("div", {
        className: cls.dialog
      }, showHeader && /*#__PURE__*/_react.default.createElement(_ModalHeader.default, {
        title: title,
        closable: closable,
        onCancel: onCancel
      }), /*#__PURE__*/_react.default.createElement(_ModalBody.default, null, children), footer && /*#__PURE__*/_react.default.createElement(_ModalFooter.default, null, footer)));
    }
  }]);
  return Modal;
}(_react.Component);

exports.default = Modal;
Modal.alert = void 0;
Modal.confirm = void 0;
Modal.defaultProps = {
  prefixCls: 'za-modal',
  visible: false,
  animationType: 'fade',
  animationDuration: 200,
  width: '70%',
  mask: true,
  maskType: 'normal',
  shape: 'radius',
  closable: false,
  maskClosable: false,
  destroy: true
};
