// zarm/lib/button/style/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _default2 = _interopRequireDefault(require("../../style/themes/default.native"));

var _default = {
  container: {
    flexDirection: 'row'
  },
  // base
  wrapperStyle: {
    alignItems: 'center',
    justifyContent: 'center',
    height: _default2.default.button_height_md,
    paddingLeft: _default2.default.button_padding_h_md,
    paddingRight: _default2.default.button_padding_h_md
  },
  textStyle: {
    fontSize: _default2.default.button_font_size_md
  },
  iconStyle: {
    marginRight: _default2.default.button_padding_h_md / 2
  },
  defaultWrapper: {
    borderWidth: 1,
    backgroundColor: _default2.default.button_default_background,
    borderColor: _default2.default.button_default_border
  },
  defaultText: {
    color: _default2.default.button_default_color
  },
  defaultActiveWrapper: {
    backgroundColor: _default2.default.button_default_active_background,
    borderColor: _default2.default.button_default_active_border
  },
  defaultActiveText: {
    color: _default2.default.button_default_active_color
  },
  defaultDisabledText: {
    color: _default2.default.button_disabled_color_ghost
  },
  defaultGhostWrapper: {
    backgroundColor: 'transparent',
    borderColor: _default2.default.button_default_ghost_border
  },
  defaultGhostText: {
    color: _default2.default.button_default_ghost_color
  },
  defaultGhostActiveWrapper: {
    backgroundColor: 'transparent',
    borderColor: _default2.default.button_default_ghost_active_border
  },
  defaultGhostActiveText: {
    color: _default2.default.button_default_ghost_active_color
  },
  primaryWrapper: {
    backgroundColor: _default2.default.button_primary_background
  },
  primaryText: {
    color: _default2.default.button_primary_color
  },
  primaryActiveWrapper: {
    backgroundColor: _default2.default.button_primary_active_background
  },
  primaryActiveText: {
    color: _default2.default.button_primary_active_color
  },
  primaryGhostWrapper: {
    borderColor: _default2.default.button_primary_ghost_border
  },
  primaryGhostText: {
    color: _default2.default.button_primary_ghost_color
  },
  primaryGhostActiveWrapper: {
    borderColor: _default2.default.button_primary_ghost_active_color
  },
  primaryGhostActiveText: {
    color: _default2.default.button_primary_ghost_active_color
  },
  dangerWrapper: {
    backgroundColor: _default2.default.button_danger_background
  },
  dangerText: {
    color: _default2.default.button_danger_color
  },
  dangerActiveWrapper: {
    backgroundColor: _default2.default.button_danger_active_background
  },
  dangerActiveText: {
    color: _default2.default.button_danger_active_color
  },
  dangerGhostWrapper: {
    borderColor: _default2.default.button_danger_ghost_border
  },
  dangerGhostText: {
    color: _default2.default.button_danger_ghost_color
  },
  dangerGhostActiveWrapper: {
    borderColor: _default2.default.button_danger_ghost_active_color
  },
  dangerGhostActiveText: {
    color: _default2.default.button_danger_ghost_active_color
  },
  // size
  mdWrapper: {
    height: _default2.default.button_height_md,
    paddingLeft: _default2.default.button_padding_h_md,
    paddingRight: _default2.default.button_padding_h_md
  },
  mdCircleWrapper: {
    width: _default2.default.button_height_md
  },
  mdText: {
    fontSize: _default2.default.button_font_size_md
  },
  mdIcon: {
    marginRight: _default2.default.button_padding_h_md / 2
  },
  lgWrapper: {
    height: _default2.default.button_height_lg,
    paddingLeft: _default2.default.button_padding_h_lg,
    paddingRight: _default2.default.button_padding_h_lg
  },
  lgCircleWrapper: {
    width: _default2.default.button_height_lg
  },
  lgText: {
    fontSize: _default2.default.button_font_size_lg
  },
  lgIcon: {
    marginRight: _default2.default.button_padding_h_lg / 2
  },
  smWrapper: {
    height: _default2.default.button_height_sm,
    paddingLeft: _default2.default.button_padding_h_sm,
    paddingRight: _default2.default.button_padding_h_sm
  },
  smCircleWrapper: {
    width: _default2.default.button_height_sm
  },
  smText: {
    fontSize: _default2.default.button_font_size_sm
  },
  smIcon: {
    marginRight: _default2.default.button_padding_h_sm / 2
  },
  xsWrapper: {
    height: _default2.default.button_height_xs,
    paddingLeft: _default2.default.button_padding_h_xs,
    paddingRight: _default2.default.button_padding_h_xs
  },
  xsCircleWrapper: {
    width: _default2.default.button_height_xs
  },
  xsText: {
    fontSize: _default2.default.button_font_size_xs
  },
  xsIcon: {
    marginRight: _default2.default.button_padding_h_xs / 2
  },
  // shape
  radiusWrapper: {
    borderRadius: _default2.default.radius_md
  },
  rectWrapper: {
    borderRadius: 0
  },
  roundWrapper: {
    borderRadius: _default2.default.radius_round
  },
  circleWrapper: {
    width: _default2.default.button_height_md,
    borderRadius: _default2.default.radius_round,
    paddingLeft: 0,
    paddingRight: 0
  },
  // ghost
  ghostWrapper: {
    borderWidth: 1,
    backgroundColor: 'transparent'
  },
  // disabled
  disabledWrapper: {
    opacity: _default2.default.button_disabled_opacity
  },
  disabledText: {
    color: _default2.default.color_text_disabled
  },
  // disabled ghost
  disabledGhostWrapper: {
    opacity: 1,
    borderColor: _default2.default.button_disabled_color_ghost
  },
  // disabled ghost text
  disabledGhostText: {
    color: _default2.default.button_disabled_color_ghost
  }
};
exports.default = _default;
