// zarm/lib/config-provider/ConfigProvider.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.getRunTimeLocale = exports.ConfigContext = exports.LocaleContext = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _setTheme = _interopRequireDefault(require("./setTheme"));

var _setPrimaryColor = _interopRequireDefault(require("./setPrimaryColor"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var defaultConfig = {
  locale: {},
  theme: 'light',
  primaryColor: '#00bc70'
};
var LocaleContext = /*#__PURE__*/(0, _react.createContext)(defaultConfig.locale);
exports.LocaleContext = LocaleContext;
var ConfigContext = /*#__PURE__*/(0, _react.createContext)(defaultConfig);
exports.ConfigContext = ConfigContext;
var runTimeLocale;

var changeRunTimeLocale = function changeRunTimeLocale(locale) {
  runTimeLocale = locale;
};

var getRunTimeLocale = function getRunTimeLocale() {
  return runTimeLocale;
};

exports.getRunTimeLocale = getRunTimeLocale;

var ConfigProvider = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(ConfigProvider, _PureComponent);

  var _super = _createSuper(ConfigProvider);

  function ConfigProvider() {
    (0, _classCallCheck2.default)(this, ConfigProvider);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(ConfigProvider, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.update();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      this.update();
    }
  }, {
    key: "update",
    value: function update() {
      var _this$props = this.props,
          locale = _this$props.locale,
          theme = _this$props.theme,
          primaryColor = _this$props.primaryColor;
      changeRunTimeLocale(locale);
      (0, _setTheme.default)(theme);
      (0, _setPrimaryColor.default)(primaryColor);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          locale = _this$props2.locale,
          theme = _this$props2.theme,
          primaryColor = _this$props2.primaryColor,
          children = _this$props2.children;
      return /*#__PURE__*/_react.default.createElement(ConfigContext.Provider, {
        value: {
          locale: locale,
          theme: theme,
          primaryColor: primaryColor
        }
      }, _react.default.Children.only(children));
    }
  }]);
  return ConfigProvider;
}(_react.PureComponent);

exports.default = ConfigProvider;
ConfigProvider.defaultProps = defaultConfig;
