// zarm/lib/action-sheet/style/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _reactNative = require("react-native");

var _default2 = _interopRequireDefault(require("../../style/themes/default.native"));

var _default = {
  wrapper: {
    margin: _default2.default.actionSheet_margin
  },
  wrapperSpacing: {
    margin: 10,
    marginTop: 0,
    borderRadius: _default2.default.radius_md,
    overflow: 'hidden'
  },
  wrapperActions: {
    backgroundColor: _default2.default.color_text_inverse
  },
  wrapperCacnel: {
    backgroundColor: _default2.default.color_text_inverse,
    marginTop: 7
  },
  actionItem: {
    height: _default2.default.actionSheet_item_height,
    lineHeight: _default2.default.actionSheet_item_height,
    fontSize: _default2.default.actionSheet_item_font_size,
    textAlign: 'center'
  },
  actionCancelItem: {
    height: _default2.default.actionSheet_item_height,
    lineHeight: _default2.default.actionSheet_item_height,
    fontSize: _default2.default.actionSheet_item_font_size,
    textAlign: 'center',
    color: _default2.default.color_text_caption
  },
  actionItemMask: {
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    height: '100%'
  },
  actionItemActive: {
    backgroundColor: _default2.default.background_active
  },
  actionItemBorder: {
    borderTopWidth: _reactNative.StyleSheet.hairlineWidth,
    borderTopColor: '#ddd'
  },
  'theme-primary': {
    color: _default2.default.theme_primary
  },
  'theme-success': {
    color: _default2.default.theme_success
  },
  'theme-warning': {
    color: _default2.default.theme_warning
  },
  'theme-danger': {
    color: _default2.default.theme_danger
  }
};
exports.default = _default;
