// zarm/lib/affix/Affix.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _events = _interopRequireDefault(require("../utils/events"));

var _dom = require("../utils/dom");

var _throttle = _interopRequireDefault(require("../utils/throttle"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var DEFAULT_SCROLL_CONTAINER = _dom.canUseDOM ? window : undefined;

var Affix = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(Affix, _PureComponent);

  var _super = _createSuper(Affix);

  function Affix() {
    var _this;

    (0, _classCallCheck2.default)(this, Affix);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.savePlaceholderNode = /*#__PURE__*/(0, _react.createRef)();
    _this.saveFixedNode = /*#__PURE__*/(0, _react.createRef)();
    _this.saveFixedNodeTop = 'offsetBottom' in _this.props ? -10000 : 10000;
    _this.state = {
      affixed: false,
      width: 0,
      height: 0
    };
    _this.onPositionUpdate = (0, _throttle.default)(function () {
      var onChange = _this.props.onChange;
      var affixed = _this.state.affixed;
      var target = _this.savePlaceholderNode.current;

      if (!target) {
        return false;
      }

      var _target$getBoundingCl = target === null || target === void 0 ? void 0 : target.getBoundingClientRect(),
          top = _target$getBoundingCl.top,
          width = _target$getBoundingCl.width,
          height = _target$getBoundingCl.height;

      _this.saveFixedNodeTop = top;
      var currentAffixed = _this.affixed;

      if (currentAffixed !== affixed) {
        _this.setState({
          affixed: currentAffixed,
          // use 'auto' when get width or height is 0
          width: width === 0 ? 'auto' : width,
          height: height === 0 ? 'auto' : height
        });

        onChange && onChange(currentAffixed);
      }
    }, 10);
    return _this;
  }

  (0, _createClass2.default)(Affix, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      // wait for ref not null
      var offsetBottom = this.props.offsetBottom;
      setTimeout(function () {
        var container = _this2.container,
            onPositionUpdate = _this2.onPositionUpdate;

        _events.default.on(container, 'scroll', onPositionUpdate);

        if (typeof offsetBottom !== 'undefined') {
          _this2.onPositionUpdate();
        }
      });
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      var _this3 = this;

      setTimeout(function () {
        var container = _this3.container,
            onPositionUpdate = _this3.onPositionUpdate;

        _events.default.off(container, 'scroll', onPositionUpdate);
      });
    }
  }, {
    key: "container",
    get: function get() {
      var scrollContainer = this.props.scrollContainer;
      var container = typeof scrollContainer === 'function' ? scrollContainer() : scrollContainer;
      return !container ? window : container;
    }
  }, {
    key: "containerRect",
    get: function get() {
      var container = this.container;

      if (!_dom.canUseDOM) {
        return {
          top: 0,
          bottom: 0
        };
      }

      return container !== window ? container.getBoundingClientRect() : {
        top: 0,
        bottom: container.innerHeight,
        width: 0,
        height: 0
      };
    }
  }, {
    key: "affixed",
    get: function get() {
      var containerRect = this.containerRect,
          saveFixedNodeTop = this.saveFixedNodeTop;
      var _this$props = this.props,
          offsetTop = _this$props.offsetTop,
          offsetBottom = _this$props.offsetBottom;

      if (typeof offsetBottom !== 'undefined' && saveFixedNodeTop + offsetBottom >= containerRect.bottom) {
        return true;
      }

      if (typeof offsetBottom === 'undefined' && typeof offsetTop !== 'undefined' && saveFixedNodeTop - offsetTop <= containerRect.top) {
        return true;
      }

      return false;
    }
  }, {
    key: "affixStyle",
    get: function get() {
      var containerRect = this.containerRect;
      var _this$props2 = this.props,
          offsetTop = _this$props2.offsetTop,
          offsetBottom = _this$props2.offsetBottom;
      var _this$state = this.state,
          width = _this$state.width,
          height = _this$state.height;

      if (this.affixed && typeof offsetBottom !== 'undefined') {
        return {
          position: 'fixed',
          bottom: offsetBottom,
          width: width,
          height: height
        };
      }

      if (this.affixed && typeof offsetTop !== 'undefined') {
        return {
          position: 'fixed',
          top: containerRect.top + offsetTop,
          width: width,
          height: height
        };
      }

      return {};
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props3 = this.props,
          prefixCls = _this$props3.prefixCls,
          className = _this$props3.className,
          children = _this$props3.children;
      var cls = (0, _classnames.default)(prefixCls, className);

      if (!this.affixed) {
        return /*#__PURE__*/_react.default.createElement("div", {
          ref: this.savePlaceholderNode
        }, children);
      }

      return /*#__PURE__*/_react.default.createElement("div", {
        ref: this.savePlaceholderNode
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: cls,
        ref: this.saveFixedNode,
        style: this.affixStyle
      }, children));
    }
  }]);
  return Affix;
}(_react.PureComponent);

exports.default = Affix;
Affix.displayName = 'Affix';
Affix.defaultProps = {
  prefixCls: 'za-affix',
  scrollContainer: DEFAULT_SCROLL_CONTAINER,
  offsetTop: 0
};
