// zarm/lib/popup/style/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _reactNative = require("react-native");

var _default2 = _interopRequireDefault(require("../../style/themes/default.native"));

var _default = {
  container: {
    flexDirection: 'row'
  },
  // base
  wrapperStyle: {
    // [@REMOVE] paddingLeft: variables.button_padding_h_md,
    // [@REMOVE] paddingRight: variables.button_padding_h_md,
    position: 'absolute',
    zIndex: _default2.default.zindex_popup // [@REMOVE] backgroundColor: '#fff',

  },
  // position
  bottomWrapper: {
    left: 0,
    width: '100%'
  },
  topWrapper: {
    left: 0,
    width: '100%',
    backgroundColor: 'rgba(0, 0 , 0, 0.7)',
    justifyContent: 'center'
  },
  leftWrapper: {
    top: 0,
    height: '100%'
  },
  rightWrapper: {
    top: 0,
    height: '100%'
  },
  invisibleWrapper: {
    height: 1,
    position: 'absolute',
    backgroundColor: 'rgba(0,0,0,0)'
  },
  topInvisible: {
    left: 0,
    top: 0,
    width: _reactNative.Dimensions.get('screen').width
  },
  bottomInvisible: {
    left: 0,
    bottom: 0,
    width: _reactNative.Dimensions.get('screen').width
  },
  leftInvisible: {
    left: 0,
    top: 0,
    height: _reactNative.Dimensions.get('screen').height
  },
  rightInvisible: {
    right: 0,
    top: 0,
    height: _reactNative.Dimensions.get('screen').height
  },
  topMask: {
    bottom: _reactNative.Dimensions.get('screen').height
  },
  bottomMask: {
    top: -_reactNative.Dimensions.get('screen').height
  },
  leftMask: {
    right: _reactNative.Dimensions.get('screen').width
  },
  rightMask: {
    left: -_reactNative.Dimensions.get('screen').width
  }
};
exports.default = _default;
