// zarm/lib/badge/style/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _default2 = _interopRequireDefault(require("../../style/themes/default.native"));

var _default = {
  textStyle: {
    flexDirection: 'row'
  },
  dotBadge: {
    width: _default2.default.badge_dot_diameter,
    height: _default2.default.badge_dot_diameter,
    borderRadius: _default2.default.badge_dot_diameter
  },
  radiusBadge: {
    borderRadius: _default2.default.badge_rect_radius,
    height: _default2.default.badge_height
  },
  roundBadge: {
    borderRadius: _default2.default.badge_dot_diameter,
    height: _default2.default.badge_height
  },
  rectBadge: {
    height: _default2.default.badge_height
  },
  circleBadge: {
    borderRadius: _default2.default.badge_dot_diameter,
    height: _default2.default.badge_height
  },
  leafBadge: {
    borderTopLeftRadius: _default2.default.badge_dot_diameter,
    borderTopRightRadius: _default2.default.badge_dot_diameter,
    borderBottomRightRadius: _default2.default.badge_dot_diameter,
    height: _default2.default.badge_height
  },
  sup: {
    position: 'absolute',
    top: _default2.default.badge_sup_top
  },
  dangerBagde: {
    backgroundColor: _default2.default.theme_danger
  },
  successBagde: {
    backgroundColor: _default2.default.theme_success
  },
  warningBagde: {
    backgroundColor: _default2.default.theme_warning
  },
  primaryBagde: {
    backgroundColor: _default2.default.theme_primary
  },
  defaultBagde: {
    backgroundColor: _default2.default.theme_default
  },
  badgeText: {
    color: _default2.default.badge_text_color,
    display: 'flex',
    fontSize: _default2.default.badge_font_size,
    lineHeight: _default2.default.badge_height,
    paddingHorizontal: _default2.default.badge_padding_h
  },
  dotText: {
    width: 0
  }
};
exports.default = _default;
