// zarm/lib/tab-bar/TabBarItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _badge = _interopRequireDefault(require("../badge"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var TabBarItem = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(TabBarItem, _PureComponent);

  var _super = _createSuper(TabBarItem);

  function TabBarItem() {
    var _this;

    (0, _classCallCheck2.default)(this, TabBarItem);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _this.change = function (value) {
      var onChange = _this.props.onChange;

      if (typeof onChange === 'function') {
        onChange(value);
      }
    };

    return _this;
  }

  (0, _createClass2.default)(TabBarItem, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          prefixCls = _this$props.prefixCls,
          title = _this$props.title,
          icon = _this$props.icon,
          badge = _this$props.badge,
          style = _this$props.style,
          itemKey = _this$props.itemKey,
          selected = _this$props.selected,
          _this$props$activeIco = _this$props.activeIcon,
          activeIcon = _this$props$activeIco === void 0 ? icon : _this$props$activeIco;
      var cls = (0, _classnames2.default)("".concat(prefixCls, "__item"), (0, _defineProperty2.default)({}, "".concat(prefixCls, "--active"), selected));

      var contentRender = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__icon")
      }, selected ? activeIcon : icon), /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__title")
      }, title));

      return /*#__PURE__*/_react.default.createElement("div", {
        className: cls,
        style: style,
        onClick: function onClick() {
          _this2.change(itemKey);
        }
      }, badge ? /*#__PURE__*/_react.default.createElement(_badge.default, badge, contentRender) : contentRender);
    }
  }]);
  return TabBarItem;
}(_react.PureComponent);

TabBarItem.defaultProps = {
  prefixCls: 'za-tab-bar'
};
var _default = TabBarItem;
exports.default = _default;
