// zarm/lib/tab-bar/TabBar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _classnames2 = _interopRequireDefault(require("classnames"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var TabBar = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(TabBar, _PureComponent);

  var _super = _createSuper(TabBar);

  function TabBar() {
    var _this;

    (0, _classCallCheck2.default)(this, TabBar);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _this.onChildChange = function (value) {
      var onChange = _this.props.onChange;

      if (typeof onChange === 'function') {
        onChange(value);
      }
    };

    _this.getSelected = function (index, itemKey) {
      var _this$props = _this.props,
          activeKey = _this$props.activeKey,
          defaultActiveKey = _this$props.defaultActiveKey;

      if (!activeKey) {
        if (!defaultActiveKey && index === 0) {
          return true;
        }

        return defaultActiveKey === itemKey;
      }

      return activeKey === itemKey;
    };

    return _this;
  }

  (0, _createClass2.default)(TabBar, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props2 = this.props,
          visible = _this$props2.visible,
          prefixCls = _this$props2.prefixCls,
          className = _this$props2.className,
          children = _this$props2.children,
          style = _this$props2.style;
      var cls = (0, _classnames2.default)(prefixCls, className, (0, _defineProperty2.default)({}, "".concat(prefixCls, "--hidden"), !visible));

      var items = _react.default.Children.map(children, function (element, index) {
        if (! /*#__PURE__*/_react.default.isValidElement(element)) return null;
        var itemKey = element.props.itemKey || index;
        return /*#__PURE__*/(0, _react.cloneElement)(element, {
          key: index,
          disabled: element.props.disabled,
          onChange: function onChange() {
            return _this2.onChildChange(itemKey);
          },
          badge: element.props.badge,
          title: element.props.title,
          icon: element.props.icon,
          itemKey: itemKey,
          style: element.props.style,
          selected: _this2.getSelected(index, itemKey)
        });
      });

      return /*#__PURE__*/_react.default.createElement("div", {
        className: cls,
        style: style
      }, items);
    }
  }]);
  return TabBar;
}(_react.PureComponent);

TabBar.Item = void 0;
TabBar.defaultProps = {
  prefixCls: 'za-tab-bar',
  visible: true
};
var _default = TabBar;
exports.default = _default;
