// zarm/lib/stepper/style/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _default2 = _interopRequireDefault(require("../../style/themes/default.native"));

var _default = {
  container: {
    flexDirection: 'row'
  },
  button: {
    alignItems: 'center',
    justifyContent: 'center',
    width: _default2.default.stepper_height_md,
    height: _default2.default.stepper_height_md,
    borderWidth: 1,
    borderColor: _default2.default.theme_primary
  },
  buttonText: {
    textAlign: 'center',
    fontSize: _default2.default.stepper_icon_font_size_md,
    color: _default2.default.theme_primary,
    marginBottom: 2
  },
  input: {
    width: _default2.default.stepper_input_width_md,
    height: _default2.default.stepper_height_md,
    padding: 0,
    marginHorizontal: 10,
    fontSize: _default2.default.stepper_input_font_size_md,
    color: _default2.default.color_text,
    textAlign: 'center'
  },
  // shape
  radiusButton: {
    borderRadius: _default2.default.radius_md
  },
  circleButton: {
    borderRadius: _default2.default.stepper_height
  },
  // disabled
  disabledButton: {
    borderColor: _default2.default.background_disabled
  },
  disabledText: {
    color: _default2.default.color_text_disabled
  }
};
exports.default = _default;
