// zarm/lib/activity-indicator/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _classnames3 = _interopRequireDefault(require("classnames"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var DIAMETER = 62;

var Circular = function Circular(props) {
  var _classnames;

  var prefixCls = props.prefixCls,
      className = props.className,
      size = props.size,
      percent = props.percent,
      strokeWidth = props.strokeWidth,
      loading = props.loading,
      style = props.style;
  var cls = (0, _classnames3.default)(className, prefixCls, (_classnames = {}, (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--").concat(size), !!size), (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--circular"), loading), _classnames));
  var half = DIAMETER / 2;
  var r = half - strokeWidth / 2;
  var round = 2 * Math.PI * r;
  var lineStyle = {
    strokeDasharray: "".concat(round * percent / 100, " ").concat(round),
    strokeWidth: strokeWidth
  };

  if (loading) {
    return /*#__PURE__*/_react.default.createElement("span", {
      className: cls,
      style: style
    }, /*#__PURE__*/_react.default.createElement("svg", {
      viewBox: "".concat(DIAMETER / 2, " ").concat(DIAMETER / 2, " ").concat(DIAMETER, " ").concat(DIAMETER)
    }, /*#__PURE__*/_react.default.createElement("circle", {
      cx: DIAMETER,
      cy: DIAMETER,
      r: r,
      fill: "none",
      style: {
        strokeWidth: strokeWidth
      }
    })));
  }

  return /*#__PURE__*/_react.default.createElement("span", {
    className: cls,
    style: style
  }, /*#__PURE__*/_react.default.createElement("svg", {
    viewBox: "0 0 ".concat(DIAMETER, " ").concat(DIAMETER)
  }, /*#__PURE__*/_react.default.createElement("circle", {
    className: "".concat(prefixCls, "__path"),
    cx: half,
    cy: half,
    r: r,
    fill: "none",
    style: {
      strokeWidth: strokeWidth
    }
  }), /*#__PURE__*/_react.default.createElement("circle", {
    className: "".concat(prefixCls, "__line"),
    cx: half,
    cy: half,
    r: r,
    fill: "none",
    style: lineStyle
  })));
};

var Spinner = function Spinner(props) {
  var prefixCls = props.prefixCls,
      className = props.className,
      size = props.size,
      style = props.style;
  var cls = (0, _classnames3.default)(prefixCls, "".concat(prefixCls, "--spinner"), className, (0, _defineProperty2.default)({}, "".concat(prefixCls, "--").concat(size), !!size));
  var spinner = [];

  for (var i = 0; i < 12; i++) {
    spinner.push( /*#__PURE__*/_react.default.createElement("div", {
      key: i
    }));
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    className: cls,
    style: style
  }, spinner);
};

var ActivityIndicator = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(ActivityIndicator, _PureComponent);

  var _super = _createSuper(ActivityIndicator);

  function ActivityIndicator() {
    (0, _classCallCheck2.default)(this, ActivityIndicator);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(ActivityIndicator, [{
    key: "render",
    value: function render() {
      var type = this.props.type;
      return type !== 'spinner' ? /*#__PURE__*/_react.default.createElement(Circular, this.props) : /*#__PURE__*/_react.default.createElement(Spinner, this.props);
    }
  }]);
  return ActivityIndicator;
}(_react.PureComponent);

exports.default = ActivityIndicator;
ActivityIndicator.defaultProps = {
  prefixCls: 'za-activity-indicator',
  strokeWidth: 5,
  percent: 20,
  type: 'circular',
  loading: true
};
