// zarm/es/alert/Alert.js
import _extends from "@babel/runtime/helpers/extends";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import React, { PureComponent } from 'react';
import classnames from 'classnames';
import Modal from '../modal';
import alertLocale from './locale/zh_CN';

var Alert = /*#__PURE__*/function (_PureComponent) {
  _inherits(Alert, _PureComponent);

  var _super = _createSuper(Alert);

  function Alert() {
    _classCallCheck(this, Alert);

    return _super.apply(this, arguments);
  }

  _createClass(Alert, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          prefixCls = _this$props.prefixCls,
          className = _this$props.className,
          content = _this$props.content,
          cancelText = _this$props.cancelText,
          onCancel = _this$props.onCancel,
          locale = _this$props.locale,
          shape = _this$props.shape,
          others = _objectWithoutProperties(_this$props, ["prefixCls", "className", "content", "cancelText", "onCancel", "locale", "shape"]);

      var cls = {
        alert: classnames(prefixCls, className, _defineProperty({}, "".concat(prefixCls, "--").concat(shape), !!shape))
      };
      return /*#__PURE__*/React.createElement("div", {
        className: cls.alert
      }, /*#__PURE__*/React.createElement(Modal, _extends({}, others, {
        footer: /*#__PURE__*/React.createElement("button", {
          type: "button",
          className: "".concat(prefixCls, "__button"),
          onClick: onCancel
        }, cancelText || locale.cancelText)
      }), content));
    }
  }]);

  return Alert;
}(PureComponent);

Alert.defaultProps = {
  prefixCls: 'za-alert',
  animationType: 'zoom',
  locale: alertLocale,
  shape: 'radius'
};
export { Alert as default };
