// zarm/lib/message/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _icons = require("@zarm-design/icons");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var Message = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(Message, _PureComponent);

  var _super = _createSuper(Message);

  function Message(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Message);
    _this = _super.call(this, props);

    _this.onClick = function (e) {
      var _this$props = _this.props,
          hasArrow = _this$props.hasArrow,
          onClick = _this$props.onClick;

      if (hasArrow && typeof onClick === 'function') {
        onClick(e);
      }
    };

    _this.onClose = function () {
      _this.setState({
        visible: false
      });
    };

    _this.state = {
      visible: true
    };
    return _this;
  }

  (0, _createClass2.default)(Message, [{
    key: "render",
    value: function render() {
      var _classnames;

      var _this$props2 = this.props,
          prefixCls = _this$props2.prefixCls,
          className = _this$props2.className,
          theme = _this$props2.theme,
          size = _this$props2.size,
          icon = _this$props2.icon,
          hasArrow = _this$props2.hasArrow,
          closable = _this$props2.closable,
          children = _this$props2.children;
      var visible = this.state.visible;
      var classes = (0, _classnames2.default)(prefixCls, className, (_classnames = {}, (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--").concat(theme), !!theme), (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--").concat(size), !!size), (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--link"), !!hasArrow), _classnames));

      var iconRender = icon && /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__icon")
      }, icon);

      var renderCloseIcon = closable && /*#__PURE__*/_react.default.createElement(_icons.Close, {
        onClick: this.onClose
      });

      var renderArrow = hasArrow && /*#__PURE__*/_react.default.createElement(_icons.ArrowRight, null);

      var noFooter = !closable && !hasArrow;
      return visible && /*#__PURE__*/_react.default.createElement("div", {
        className: classes,
        onClick: this.onClick
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__header")
      }, iconRender), /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__body")
      }, children), !noFooter && /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__footer")
      }, renderArrow, renderCloseIcon));
    }
  }]);
  return Message;
}(_react.PureComponent);

exports.default = Message;
Message.defaultProps = {
  prefixCls: 'za-message',
  theme: 'primary',
  hasArrow: false,
  closable: false
};
