// zarm/lib/nav-bar/index.native.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _index = _interopRequireDefault(require("./style/index.native"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var navBarStyles = _reactNative.StyleSheet.create(_index.default);

var PADDING_TOP = _reactNative.Platform.OS === 'ios' ? 20 : 0;

var NavBar = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(NavBar, _PureComponent);

  var _super = _createSuper(NavBar);

  function NavBar() {
    (0, _classCallCheck2.default)(this, NavBar);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(NavBar, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          styles = _this$props.styles,
          style = _this$props.style,
          left = _this$props.left,
          right = _this$props.right;
      var navBarWrapperStyle = [styles.navBarWrapper, style, {
        paddingTop: PADDING_TOP
      }];

      var leftBtn = /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, null, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
        style: [styles.btn]
      }, left));

      var rightBtn = right ? /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, null, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
        style: [styles.btn, styles.rightBtn]
      }, right)) : /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
        style: styles.btn
      });
      return /*#__PURE__*/_react.default.createElement(_reactNative.View, null, /*#__PURE__*/_react.default.createElement(_reactNative.StatusBar, {
        barStyle: "default"
      }), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: navBarWrapperStyle
      }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: styles.navBarBox
      }, leftBtn, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: styles.titleWrapper
      }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
        style: styles.titleText
      }, "Photos")), rightBtn)));
    }
  }]);
  return NavBar;
}(_react.PureComponent);

exports.default = NavBar;
NavBar.defaultProps = {
  styles: navBarStyles
  /* leftText: '返回',
  rightText: '确定',
  showRight: true, */

};
