// zarm/es/file-picker/utils/handleFileInfo.js
import getFileDetail from './getFileDetail';
import createThumbnail from './createThumbnail';

/**
 * handle single file and get file info
 *
 * @param file
 * @param quality
 * @param callback
 */
export default function handleFileInfo(_ref, callback) {
  var file = _ref.file,
      quality = _ref.quality;

  var _getFileDetail = getFileDetail(file),
      fileSize = _getFileDetail.fileSize,
      fileType = _getFileDetail.fileType,
      isPic = _getFileDetail.isPic,
      fileName = _getFileDetail.fileName;

  var fileDetail = {
    file: file,
    fileType: fileType,
    fileSize: fileSize,
    fileName: fileName,
    thumbnail: ''
  };

  var callbackFunc = function callbackFunc(url) {
    fileDetail.thumbnail = url;
    callback(fileDetail);
  };

  if (isPic) {
    createThumbnail({
      file: file,
      quality: quality,
      fileType: fileType
    }).then(function (url) {
      return callbackFunc(url);
    }).catch(function (e) {
      return console.error(e);
    });
  } else {
    callback(fileDetail);
  }
}
