// zarm/lib/date-select/DateSelect.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _format = _interopRequireDefault(require("../date-picker-view/utils/format"));

var _datePicker = _interopRequireDefault(require("../date-picker"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var DateSelect = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(DateSelect, _PureComponent);

  var _super = _createSuper(DateSelect);

  function DateSelect(props) {
    var _this;

    (0, _classCallCheck2.default)(this, DateSelect);
    _this = _super.call(this, props);

    _this.handleClick = function () {
      var disabled = _this.props.disabled;

      if (disabled) {
        return false;
      }

      _this.setState({
        visible: true
      });
    };

    _this.onChange = function (selected) {
      var onChange = _this.props.onChange;

      if (typeof onChange === 'function') {
        onChange(selected);
      }
    };

    _this.onOk = function (selected) {
      var onOk = _this.props.onOk;

      _this.setState({
        visible: false,
        selectValue: selected
      });

      if (typeof onOk === 'function') {
        onOk(selected);
      }
    };

    _this.onCancel = function () {
      var onCancel = _this.props.onCancel;

      _this.setState({
        visible: false
      });

      if (typeof onCancel === 'function') {
        onCancel();
      }
    };

    _this.state = {
      visible: false,
      selectValue: props.value
    };
    return _this;
  }

  (0, _createClass2.default)(DateSelect, [{
    key: "render",
    value: function render() {
      var _classnames;

      var _this$props = this.props,
          prefixCls = _this$props.prefixCls,
          className = _this$props.className,
          placeholder = _this$props.placeholder,
          disabled = _this$props.disabled,
          onChange = _this$props.onChange,
          locale = _this$props.locale,
          value = _this$props.value,
          hasArrow = _this$props.hasArrow,
          others = (0, _objectWithoutProperties2.default)(_this$props, ["prefixCls", "className", "placeholder", "disabled", "onChange", "locale", "value", "hasArrow"]);
      var _this$state = this.state,
          visible = _this$state.visible,
          selectValue = _this$state.selectValue;
      var cls = (0, _classnames2.default)(prefixCls, (_classnames = {}, (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--placeholder"), !selectValue), (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--disabled"), disabled), (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--arrow"), hasArrow), _classnames));
      return /*#__PURE__*/_react.default.createElement("div", {
        className: cls,
        onClick: this.handleClick
      }, /*#__PURE__*/_react.default.createElement("input", {
        type: "hidden",
        value: (0, _format.default)(this, selectValue)
      }), /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__input")
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__value")
      }, (0, _format.default)(this, selectValue) || placeholder || locale.placeholder)), /*#__PURE__*/_react.default.createElement(_datePicker.default, (0, _extends2.default)({}, others, {
        className: className,
        visible: visible,
        value: selectValue,
        onOk: this.onOk,
        onCancel: this.onCancel
      })));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(props) {
      return {
        selectValue: props.value
      };
    }
  }]);
  return DateSelect;
}(_react.PureComponent);

exports.default = DateSelect;
DateSelect.defaultProps = {
  prefixCls: 'za-date-select',
  mode: 'date',
  disabled: false,
  minuteStep: 1,
  valueMember: 'value',
  hasArrow: true,
  onCancel: function onCancel() {}
};
