// zarm/lib/select/Select.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _isEqual = _interopRequireDefault(require("lodash/isEqual"));

var _picker = _interopRequireDefault(require("../picker"));

var _parseProps = _interopRequireDefault(require("../picker-view/utils/parseProps"));

var _removeFnFromProps = _interopRequireDefault(require("../picker-view/utils/removeFnFromProps"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var Select = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(Select, _PureComponent);

  var _super = _createSuper(Select);

  function Select(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Select);
    _this = _super.call(this, props);

    _this.handleClick = function () {
      var disabled = _this.props.disabled;

      if (disabled) {
        return false;
      }

      _this.setState({
        visible: true
      });
    };

    _this.onChange = function (selected) {
      var onChange = _this.props.onChange;

      if (typeof onChange === 'function') {
        onChange(selected);
      }
    };

    _this.onOk = function (selected) {
      var onOk = _this.props.onOk;

      _this.setState({
        selectValue: selected,
        visible: false
      }, function () {
        if (typeof onOk === 'function') {
          onOk(selected);
        }
      });
    };

    _this.onCancel = function () {
      var onCancel = _this.props.onCancel;

      _this.setState({
        visible: false
      });

      if (typeof onCancel === 'function') {
        onCancel();
      }
    };

    _this.state = {
      visible: false,
      selectValue: _parseProps.default.isValueValid(props.defaultValue || props.value) && _parseProps.default.getSource(props).objValue
    };
    return _this;
  }

  (0, _createClass2.default)(Select, [{
    key: "render",
    value: function render() {
      var _classnames;

      var _this$props = this.props,
          prefixCls = _this$props.prefixCls,
          placeholder = _this$props.placeholder,
          className = _this$props.className,
          disabled = _this$props.disabled,
          displayRender = _this$props.displayRender,
          locale = _this$props.locale,
          value = _this$props.value,
          hasArrow = _this$props.hasArrow,
          others = (0, _objectWithoutProperties2.default)(_this$props, ["prefixCls", "placeholder", "className", "disabled", "displayRender", "locale", "value", "hasArrow"]);
      var _this$state = this.state,
          visible = _this$state.visible,
          _this$state$selectVal = _this$state.selectValue,
          selectValue = _this$state$selectVal === void 0 ? [] : _this$state$selectVal;
      var cls = (0, _classnames2.default)(prefixCls, (_classnames = {}, (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--placeholder"), !selectValue.length), (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--disabled"), disabled), (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--arrow"), hasArrow), _classnames));
      return /*#__PURE__*/_react.default.createElement("div", {
        className: cls,
        onClick: this.handleClick
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__input")
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__value")
      }, selectValue.length && displayRender(selectValue || []) || placeholder || locale.placeholder)), /*#__PURE__*/_react.default.createElement(_picker.default, (0, _extends2.default)({}, others, {
        className: className,
        visible: visible,
        value: value,
        onOk: this.onOk,
        onChange: this.onChange,
        onCancel: this.onCancel
      })));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(props, state) {
      if (!(0, _isEqual.default)((0, _removeFnFromProps.default)(props, ['onOk', 'onCancel', 'onChange']), (0, _removeFnFromProps.default)(state.prevProps, ['onOk', 'onCancel', 'onChange']))) {
        return {
          prevProps: props,
          selectValue: _parseProps.default.isValueValid(props.defaultValue || props.value) && _parseProps.default.getSource(props).objValue
        };
      }

      return null;
    }
  }]);
  return Select;
}(_react.PureComponent);

exports.default = Select;
Select.defaultProps = {
  prefixCls: 'za-select',
  dataSource: [],
  valueMember: 'value',
  itemRender: function itemRender(data) {
    return data && data.label;
  },
  cols: Infinity,
  hasArrow: true,
  maskClosable: true,
  displayRender: function displayRender(selected) {
    return selected.map(function (item) {
      return item && item.label;
    });
  },
  onClick: function onClick() {}
};
