// zarm/lib/message/style/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _color = _interopRequireDefault(require("color"));

var _default2 = _interopRequireDefault(require("../../style/themes/default.native"));

var _default = {
  wrapper: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    minHeight: _default2.default.message_min_height_md,
    paddingHorizontal: _default2.default.message_padding_h_md,
    paddingVertical: _default2.default.message_padding_v_md,
    backgroundColor: _default2.default.message_theme_bg_primary
  },
  lgWrapper: {
    minHeight: _default2.default.message_min_height_lg,
    paddingHorizontal: _default2.default.message_padding_h_lg,
    paddingVertical: _default2.default.message_padding_v_lg
  },
  primaryWrapper: {
    backgroundColor: "".concat((0, _color.default)(_default2.default.theme_primary).alpha(0.1))
  },
  successWrapper: {
    backgroundColor: "".concat((0, _color.default)(_default2.default.theme_success).alpha(0.1))
  },
  warningWrapper: {
    backgroundColor: "".concat((0, _color.default)(_default2.default.theme_warning).alpha(0.1))
  },
  dangerWrapper: {
    backgroundColor: "".concat((0, _color.default)(_default2.default.theme_danger).alpha(0.1))
  },
  headerStyle: {},
  bodyStyle: {
    flex: 1,
    flexDirection: 'row'
  },
  footerStyle: {
    flexDirection: 'row',
    paddingLeft: 10
  },
  mdTextStyle: {
    fontSize: _default2.default.message_font_size_md
  },
  lgTextStyle: {
    fontSize: _default2.default.message_font_size_lg
  },
  primaryTextStyle: {
    color: _default2.default.theme_primary
  },
  successTextStyle: {
    color: _default2.default.theme_success
  },
  warningTextStyle: {
    color: _default2.default.theme_warning
  },
  dangerTextStyle: {
    color: _default2.default.theme_danger
  },
  // close icon
  closeIconWrapperStyle: {
    position: 'relative',
    alignItems: 'center',
    justifyContent: 'center',
    width: 14,
    height: 20
  },
  closeIconStyle: {
    position: 'absolute',
    height: 1
  },
  // close icon size
  mdCloseIconStyle: {
    width: _default2.default.message_close_length_md
  },
  lgCloseIconStyle: {
    width: _default2.default.message_close_length_lg
  },
  // close icon theme
  primaryCloseIconStyle: {
    backgroundColor: _default2.default.theme_primary
  },
  successCloseIconStyle: {
    backgroundColor: _default2.default.theme_success
  },
  warningCloseIconStyle: {
    backgroundColor: _default2.default.theme_warning
  },
  dangerCloseIconStyle: {
    backgroundColor: _default2.default.theme_danger
  },
  // close icon type
  closeIconLeft: {
    transform: [{
      rotate: '45deg'
    }]
  },
  closeIconRight: {
    transform: [{
      rotate: '135deg'
    }]
  },
  // arrow
  arrowWrapperStyle: {
    position: 'relative',
    alignItems: 'center',
    justifyContent: 'center',
    width: 14,
    height: 20
  },
  arrowStyle: {
    position: 'absolute',
    borderTopWidth: _default2.default.message_arrow_border_width,
    borderRightWidth: _default2.default.message_arrow_border_width,
    transform: [{
      rotate: '45deg'
    }]
  },
  // arrow size
  mdArrowStyle: {
    width: _default2.default.message_arrow_length_md,
    height: _default2.default.message_arrow_length_md
  },
  lgArrowStyle: {
    width: _default2.default.message_arrow_length_lg,
    height: _default2.default.message_arrow_length_lg
  },
  // arrow theme
  primaryArrowStyle: {
    borderTopColor: _default2.default.theme_primary,
    borderRightColor: _default2.default.theme_primary
  },
  successArrowStyle: {
    borderTopColor: _default2.default.theme_success,
    borderRightColor: _default2.default.theme_success
  },
  warningArrowStyle: {
    borderTopColor: _default2.default.theme_warning,
    borderRightColor: _default2.default.theme_warning
  },
  dangerArrowStyle: {
    borderTopColor: _default2.default.theme_danger,
    borderRightColor: _default2.default.theme_danger
  }
};
exports.default = _default;
