// zarm/lib/file-picker/utils/handleFileInfo.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = handleFileInfo;

var _getFileDetail2 = _interopRequireDefault(require("./getFileDetail"));

var _createThumbnail = _interopRequireDefault(require("./createThumbnail"));

/**
 * handle single file and get file info
 *
 * @param file
 * @param quality
 * @param callback
 */
function handleFileInfo(_ref, callback) {
  var file = _ref.file,
      quality = _ref.quality;

  var _getFileDetail = (0, _getFileDetail2.default)(file),
      fileSize = _getFileDetail.fileSize,
      fileType = _getFileDetail.fileType,
      isPic = _getFileDetail.isPic,
      fileName = _getFileDetail.fileName;

  var fileDetail = {
    file: file,
    fileType: fileType,
    fileSize: fileSize,
    fileName: fileName,
    thumbnail: ''
  };

  var callbackFunc = function callbackFunc(url) {
    fileDetail.thumbnail = url;
    callback(fileDetail);
  };

  if (isPic) {
    (0, _createThumbnail.default)({
      file: file,
      quality: quality,
      fileType: fileType
    }).then(function (url) {
      return callbackFunc(url);
    }).catch(function (e) {
      return console.error(e);
    });
  } else {
    callback(fileDetail);
  }
}
