// zarm/lib/input/InputNumber.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _icons = require("@zarm-design/icons");

var _events = _interopRequireDefault(require("../utils/events"));

var _keyboardPicker = _interopRequireDefault(require("../keyboard-picker"));

var _utils = require("./utils");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var InputNumber = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(InputNumber, _Component);

  var _super = _createSuper(InputNumber);

  function InputNumber(props) {
    var _this;

    (0, _classCallCheck2.default)(this, InputNumber);
    _this = _super.call(this, props);
    _this.content = void 0;
    _this.picker = void 0;
    _this.container = void 0;

    _this.addBlurListener = function () {
      _events.default.on(document.body, 'click', _this.onMaskClick);
    };

    _this.removeBlurListener = function () {
      _events.default.off(document.body, 'click', _this.onMaskClick);
    };

    _this.onMaskClick = function (e) {
      var clsRegExp = new RegExp("(^|\\s)".concat(_this.picker.props.prefixCls, "(\\s|$)"), 'g');

      if (!_this.state.visible || _this.state.focused) {
        return;
      }

      var cNode = function (node) {
        while (node.parentNode && node.parentNode !== document.body) {
          if (node === _this.picker || node === _this.container || clsRegExp.test(node.className)) {
            return node;
          }

          node = node.parentNode;
        }
      }(e.target);

      if (!cNode) {
        _this.onBlur();
      }
    };

    _this.onKeyClick = function (key) {
      if (['close', 'ok'].indexOf(key) > -1) {
        _this.onBlur();

        return;
      }

      var value = _this.state.value;
      var newValue = key === 'delete' ? String(value).slice(0, String(value).length - 1) : value + key;

      if (!('value' in _this.props)) {
        _this.setState({
          value: newValue
        });
      }

      var onChange = _this.props.onChange;

      if (typeof onChange === 'function') {
        onChange(newValue);
      }
    };

    _this.onFocus = function () {
      var _this$props = _this.props,
          disabled = _this$props.disabled,
          readOnly = _this$props.readOnly,
          onFocus = _this$props.onFocus;
      var _this$state = _this.state,
          visible = _this$state.visible,
          value = _this$state.value;

      if (disabled || readOnly || visible) {
        return;
      } // 定位到文本尾部


      _this.setState({
        visible: true
      });

      if (typeof onFocus === 'function') {
        onFocus(value);
      }
    };

    _this.onBlur = function () {
      var _this$state2 = _this.state,
          visible = _this$state2.visible,
          value = _this$state2.value;

      if (!visible) {
        return;
      } // 定位到文本首部


      _this.setState({
        visible: false
      });

      var onBlur = _this.props.onBlur;

      if (typeof onBlur === 'function') {
        onBlur(value);
      }
    };

    _this.onClear = function () {
      var _this$props2 = _this.props,
          onChange = _this$props2.onChange,
          onClear = _this$props2.onClear;

      _this.setState({
        value: ''
      }, _this.onFocus);

      if (onClear) {
        onClear('');
      }

      if (typeof onChange === 'function') {
        onChange('');
      }
    };

    _this.scrollToStart = function () {
      _this.content.scrollLeft = 0;
    };

    _this.scrollToEnd = function () {
      _this.content.scrollLeft = _this.content.scrollWidth;
    };

    _this.focus = function () {
      _this.removeBlurListener();

      _this.onFocus();

      setTimeout(function () {
        _this.addBlurListener();
      }, 50);
    };

    _this.blur = function () {
      _this.onBlur();
    };

    _this.state = {
      value: (0, _utils.getValue)(props),
      visible: props.focused || false,
      focused: false
    };
    return _this;
  }

  (0, _createClass2.default)(InputNumber, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this$props3 = this.props,
          autoFocus = _this$props3.autoFocus,
          focused = _this$props3.focused;
      this.addBlurListener();

      if (autoFocus || focused) {
        this.onFocus();
      }
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      var visible = this.state.visible;

      if (visible) {
        this.scrollToEnd();
      } else {
        this.scrollToStart();
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.removeBlurListener();
    }
  }, {
    key: "render",
    value: function render() {
      var _classnames,
          _this2 = this;

      var _this$props4 = this.props,
          prefixCls = _this$props4.prefixCls,
          className = _this$props4.className,
          type = _this$props4.type,
          clearable = _this$props4.clearable,
          disabled = _this$props4.disabled,
          readOnly = _this$props4.readOnly,
          placeholder = _this$props4.placeholder;
      var _this$state3 = this.state,
          visible = _this$state3.visible,
          value = _this$state3.value;
      var showClearIcon = clearable && 'value' in this.props && value.length > 0 && 'onChange' in this.props;
      var cls = (0, _classnames2.default)(prefixCls, "".concat(prefixCls, "--").concat(type), className, (_classnames = {}, (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--disabled"), disabled), (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--focus"), visible), (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--clearable"), showClearIcon), (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--readonly"), readOnly), _classnames));

      var renderInput = /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__content")
      }, (value === undefined || value === '') && !readOnly && /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__placeholder")
      }, placeholder), /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__virtual-input"),
        ref: function ref(ele) {
          _this2.content = ele;
        }
      }, value), /*#__PURE__*/_react.default.createElement("input", {
        type: "hidden",
        value: value,
        disabled: disabled
      }), /*#__PURE__*/_react.default.createElement(_keyboardPicker.default, {
        ref: function ref(ele) {
          _this2.picker = ele;
        },
        visible: visible,
        type: type,
        onKeyClick: this.onKeyClick
      }));

      var renderText = /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__content")
      }, value);

      var renderClearIcon = showClearIcon && /*#__PURE__*/_react.default.createElement(_icons.CloseCircleFill, {
        className: "".concat(prefixCls, "__clear"),
        onClick: function onClick(e) {
          e.stopPropagation();

          _this2.onClear();
        }
      });

      return /*#__PURE__*/_react.default.createElement("div", {
        className: cls,
        onClick: this.onFocus,
        ref: function ref(ele) {
          _this2.container = ele;
        }
      }, readOnly ? renderText : renderInput, renderClearIcon);
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, state) {
      if ('value' in nextProps && nextProps.value !== state.prevValue) {
        return {
          value: (0, _utils.getValue)(nextProps),
          prevValue: (0, _utils.getValue)(nextProps)
        };
      }

      return null;
    }
  }]);
  return InputNumber;
}(_react.Component);

exports.default = InputNumber;
InputNumber.defaultProps = {
  prefixCls: 'za-input',
  disabled: false,
  clearable: true
};
