// zarm/lib/collapse/style/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _reactNative = require("react-native");

var _default2 = _interopRequireDefault(require("../../style/themes/default.native"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var border = {
  borderBottomWidth: _reactNative.StyleSheet.hairlineWidth,
  // borderBottomWidth: 1,
  borderBottomColor: _default2.default.theme_default,
  borderStyle: 'solid'
};
var _default = {
  container: {
    flexDirection: 'column',
    backgroundColor: '#FFF'
  },
  // base
  itemWrapperStyle: {
    width: '100%',
    flexDirection: 'column',
    overflow: 'hidden',
    backgroundColor: '#FFF'
  },
  titleWrapperStyle: {
    backgroundColor: '#FFF',
    height: _default2.default.collapse_height
  },
  titleInnerStyle: _objectSpread({
    height: '100%',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    backgroundColor: '#FFF',
    // 解决在android下overflow属性无效（使用背景色掩盖文字）
    marginLeft: _default2.default.collapse_padding,
    paddingRight: _default2.default.collapse_padding
  }, border),
  titleTextStyle: {
    fontSize: 15,
    color: '#464646'
  },
  titleTextDisabledStyle: {
    color: '#999'
  },
  titleArrowStyle: {
    borderRightWidth: 2,
    borderBottomWidth: 2,
    borderStyle: 'solid',
    borderColor: _default2.default.collapse_arrow_color,
    width: 10,
    height: 10,
    position: 'relative'
  },
  titleArrowDisabledStyle: {
    borderColor: _default2.default.collapse_arrow_disabled_color
  },
  itemBodyStyle: _objectSpread({
    backgroundColor: '#FFF',
    marginLeft: _default2.default.collapse_padding,
    paddingRight: _default2.default.collapse_padding,
    paddingTop: _default2.default.collapse_padding,
    paddingBottom: _default2.default.collapse_padding
  }, border)
};
exports.default = _default;
