// zarm/lib/cell/style/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _reactNative = require("react-native");

var _default2 = _interopRequireDefault(require("../../style/themes/default.native"));

var _default = {
  wrapperStyle: {
    flex: 1,
    flexDirection: 'column',
    minHeight: _default2.default.cell_height,
    backgroundColor: _default2.default.cell_background
  },
  activeWrapper: {
    backgroundColor: _default2.default.background_active
  },
  innerStyle: {
    flex: 1,
    flexDirection: 'column',
    marginTop: -1,
    paddingHorizontal: _default2.default.cell_padding_h
  },
  lineContainerStyle: {
    paddingLeft: _default2.default.cell_line_padding_left
  },
  lineStyle: {
    borderTopWidth: _reactNative.StyleSheet.hairlineWidth,
    borderTopColor: _default2.default.border_color
  },
  contentStyle: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center'
  },
  bodyStyle: {
    flex: 1,
    flexDirection: 'row'
  },
  arrowStyle: {
    width: _default2.default.cell_arrow_length,
    height: _default2.default.cell_arrow_length,
    marginLeft: 5,
    borderTopWidth: _default2.default.cell_arrow_border_width,
    borderTopColor: _default2.default.cell_arrow_color,
    borderRightWidth: _default2.default.cell_arrow_border_width,
    borderRightColor: _default2.default.cell_arrow_color,
    transform: [{
      rotate: '45deg'
    }]
  },
  iconStyle: {
    marginRight: _default2.default.padding_h_md
  },
  descriptionTextStyle: {
    color: _default2.default.cell_description_color,
    fontSize: _default2.default.cell_description_font_size
  },
  titleViewStyle: {
    flex: 1,
    paddingVertical: _default2.default.cell_padding_v
  },
  labelTitleViewStyle: {
    flex: 0,
    width: _default2.default.cell_label_width
  },
  titleTextStyle: {
    color: _default2.default.cell_title_color,
    fontSize: _default2.default.cell_title_font_size,
    lineHeight: _default2.default.cell_title_line_height
  },
  helpViewStyle: {
    paddingHorizontal: _default2.default.cell_padding_h,
    paddingTop: 0,
    paddingBottom: 10
  },
  childrenViewStyle: {
    flex: 1,
    flexDirection: 'row',
    paddingVertical: _default2.default.cell_padding_v,
    paddingHorizontal: 0
  }
};
exports.default = _default;
