// zarm/lib/image-preview/ImagePreview.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _popup = _interopRequireDefault(require("../popup"));

var _carousel = _interopRequireDefault(require("../carousel"));

var _pinchZoom = _interopRequireDefault(require("../pinch-zoom"));

var _activityIndicator = _interopRequireDefault(require("../activity-indicator"));

var _loadStatus = _interopRequireDefault(require("./utils/loadStatus"));

var _formatImages = _interopRequireDefault(require("./utils/formatImages"));

var _showOriginButton = _interopRequireDefault(require("./utils/showOriginButton"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var parseState = function parseState(props) {
  var visible = props.visible,
      images = props.images;
  return {
    visible: visible,
    images: (0, _formatImages.default)(images)
  };
};

var ImagePreview = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(ImagePreview, _Component);

  var _super = _createSuper(ImagePreview);

  function ImagePreview() {
    var _this;

    (0, _classCallCheck2.default)(this, ImagePreview);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.doubleClickTimer = void 0;
    _this.touchStartTime = void 0;
    _this.moving = void 0;
    _this.state = parseState(_this.props);

    _this.onChange = function (index) {
      var onChange = _this.props.onChange;

      _this.setState({
        currentIndex: index
      }, function () {
        if (typeof onChange === 'function') {
          onChange(index);
        }
      });
    };

    _this.close = function () {
      if (_this.moving) {
        return false;
      }

      var onClose = _this.props.onClose;

      if (typeof onClose === 'function') {
        onClose();
      }
    };

    _this.loadOrigin = function () {
      var _this$state = _this.state,
          _this$state$currentIn = _this$state.currentIndex,
          currentIndex = _this$state$currentIn === void 0 ? 0 : _this$state$currentIn,
          images = _this$state.images;
      var _images$currentIndex = images[currentIndex],
          originUrl = _images$currentIndex.originUrl,
          loaded = _images$currentIndex.loaded;

      if (loaded !== _loadStatus.default.before || !originUrl) {
        return;
      }

      images[currentIndex].loaded = _loadStatus.default.start;

      _this.setState({
        images: images
      });

      var img = new Image();

      img.onload = function () {
        images[currentIndex].loaded = _loadStatus.default.end;
        images[currentIndex].url = originUrl;

        _this.setState({
          images: images
        });

        setTimeout(function () {
          images[currentIndex].loaded = _loadStatus.default.after;

          _this.setState({
            images: images
          });
        }, 1500);
      };

      img.src = originUrl;
    };

    _this.onWrapperTouchStart = function () {
      _this.touchStartTime = Date.now();
    };

    _this.onWrapperTouchEnd = function () {
      var deltaTime = Date.now() - _this.touchStartTime;

      var onClose = _this.props.onClose; // prevent long tap to close component

      if (deltaTime < 300) {
        if (!_this.doubleClickTimer && !_this.moving) {
          _this.doubleClickTimer = setTimeout(function () {
            _this.doubleClickTimer = null;

            if (typeof onClose === 'function') {
              onClose();
            }
          }, 300);
        } else {
          _this.doubleClickTimer && clearTimeout(_this.doubleClickTimer);
          _this.doubleClickTimer = null;
        }
      }

      _this.moving = false;
    };

    _this.onWrapperTouchMove = function () {
      if (_this.touchStartTime) {
        _this.moving = true;
      }
    };

    _this.onWrapperMouseDown = function () {
      _this.touchStartTime = Date.now();
    };

    _this.onWrapperMouseUp = function () {
      setTimeout(function () {
        _this.moving = false;
      }, 0);
      _this.touchStartTime = 0;
    };

    _this.renderImages = function () {
      var _this$props = _this.props,
          prefixCls = _this$props.prefixCls,
          minScale = _this$props.minScale,
          maxScale = _this$props.maxScale;
      var images = _this.state.images;
      return images.map(function (item, i) {
        return /*#__PURE__*/_react.default.createElement("div", {
          className: "".concat(prefixCls, "__item"),
          key: +i
        }, /*#__PURE__*/_react.default.createElement(_pinchZoom.default, {
          className: "".concat(prefixCls, "__item__img"),
          minScale: minScale,
          maxScale: maxScale
        }, /*#__PURE__*/_react.default.createElement("img", {
          src: item.url,
          alt: "",
          draggable: false
        })));
      });
    };

    return _this;
  }

  (0, _createClass2.default)(ImagePreview, [{
    key: "renderPagination",
    value: function renderPagination() {
      var _this$props2 = this.props,
          prefixCls = _this$props2.prefixCls,
          showPagination = _this$props2.showPagination;
      var _this$state2 = this.state,
          _this$state2$currentI = _this$state2.currentIndex,
          currentIndex = _this$state2$currentI === void 0 ? 0 : _this$state2$currentI,
          visible = _this$state2.visible,
          images = _this$state2.images;

      if (visible && showPagination && images && images.length > 1) {
        return /*#__PURE__*/_react.default.createElement("div", {
          className: "".concat(prefixCls, "__index")
        }, currentIndex + 1, " / ", images.length);
      }

      return null;
    }
  }, {
    key: "renderOriginButton",
    value: function renderOriginButton() {
      var _this$state3 = this.state,
          images = _this$state3.images,
          _this$state3$currentI = _this$state3.currentIndex,
          currentIndex = _this$state3$currentI === void 0 ? 0 : _this$state3$currentI;
      if (images.length === 0) return;
      var _this$props3 = this.props,
          prefixCls = _this$props3.prefixCls,
          locale = _this$props3.locale,
          activeIndex = _this$props3.activeIndex;
      var loaded = images[currentIndex].loaded;

      if (loaded && (0, _showOriginButton.default)(images, activeIndex) && loaded !== _loadStatus.default.after) {
        return /*#__PURE__*/_react.default.createElement("button", {
          className: "".concat(prefixCls, "__origin__button"),
          onClick: this.loadOrigin
        }, loaded === _loadStatus.default.start && /*#__PURE__*/_react.default.createElement(_activityIndicator.default, {
          className: "".concat(prefixCls, "__loading"),
          type: "spinner"
        }), locale && locale[loaded]);
      }

      return null;
    }
  }, {
    key: "render",
    value: function render() {
      var prefixCls = this.props.prefixCls;
      var _this$state4 = this.state,
          _this$state4$currentI = _this$state4.currentIndex,
          currentIndex = _this$state4$currentI === void 0 ? 0 : _this$state4$currentI,
          visible = _this$state4.visible,
          images = _this$state4.images;
      return /*#__PURE__*/_react.default.createElement(_popup.default, {
        direction: "center",
        visible: visible,
        className: prefixCls
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__content"),
        onTouchStart: this.onWrapperTouchStart,
        onTouchEnd: this.onWrapperTouchEnd,
        onTouchCancel: this.onWrapperTouchEnd,
        onTouchMove: this.onWrapperTouchMove,
        onMouseDown: this.onWrapperMouseDown,
        onMouseMove: this.onWrapperTouchMove,
        onMouseUp: this.onWrapperMouseUp,
        onClick: this.close
      }, visible && (images !== null && images !== void 0 && images.length ? /*#__PURE__*/_react.default.createElement(_carousel.default, {
        showPagination: false,
        onChange: this.onChange,
        activeIndex: currentIndex
      }, this.renderImages()) : /*#__PURE__*/_react.default.createElement(_activityIndicator.default, {
        className: "".concat(prefixCls, "__loading"),
        type: "spinner",
        size: "lg"
      }))), /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__footer")
      }, this.renderOriginButton(), this.renderPagination()));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, state) {
      if ('visible' in nextProps && nextProps.visible !== state.prevVisible || 'activeIndex' in nextProps && nextProps.activeIndex !== state.prevActiveIndex || 'images' in nextProps && nextProps.images !== state.prevImages) {
        return {
          visible: nextProps.visible,
          activeIndex: nextProps.activeIndex,
          currentIndex: nextProps.activeIndex,
          images: (0, _formatImages.default)(nextProps.images),
          prevVisible: nextProps.visible,
          prevActiveIndex: nextProps.activeIndex,
          prevImages: nextProps.images
        };
      }

      return null;
    }
  }]);
  return ImagePreview;
}(_react.Component);

exports.default = ImagePreview;
ImagePreview.defaultProps = {
  prefixCls: 'za-image-preview',
  activeIndex: 0,
  showPagination: true,
  visible: false
};
