// zarm/es/nav-bar/index.native.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

/**
 * 目前用于 modal 的导航，提供简单的返回、确定
 * 一般页面跳转使用 react-navigation
 *
 * showRight: 是否展示右边按钮，默认展示
 * leftText：左边按钮文案
 * rightText： 右边按钮文案
 * onClickLeft： 左边按钮点击事件
 * onClickRight： 右边按钮点击事件
 */
import React, { PureComponent } from 'react';
import { StyleSheet, View, TouchableOpacity, Text, StatusBar, Platform } from 'react-native';
import navBarStyle from './style/index.native';
var navBarStyles = StyleSheet.create(navBarStyle);
var PADDING_TOP = Platform.OS === 'ios' ? 20 : 0;

var NavBar = /*#__PURE__*/function (_PureComponent) {
  _inherits(NavBar, _PureComponent);

  var _super = _createSuper(NavBar);

  function NavBar() {
    _classCallCheck(this, NavBar);

    return _super.apply(this, arguments);
  }

  _createClass(NavBar, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          styles = _this$props.styles,
          style = _this$props.style,
          left = _this$props.left,
          right = _this$props.right;
      var navBarWrapperStyle = [styles.navBarWrapper, style, {
        paddingTop: PADDING_TOP
      }];
      var leftBtn = /*#__PURE__*/React.createElement(TouchableOpacity, null, /*#__PURE__*/React.createElement(Text, {
        style: [styles.btn]
      }, left));
      var rightBtn = right ? /*#__PURE__*/React.createElement(TouchableOpacity, null, /*#__PURE__*/React.createElement(Text, {
        style: [styles.btn, styles.rightBtn]
      }, right)) : /*#__PURE__*/React.createElement(Text, {
        style: styles.btn
      });
      return /*#__PURE__*/React.createElement(View, null, /*#__PURE__*/React.createElement(StatusBar, {
        barStyle: "default"
      }), /*#__PURE__*/React.createElement(View, {
        style: navBarWrapperStyle
      }, /*#__PURE__*/React.createElement(View, {
        style: styles.navBarBox
      }, leftBtn, /*#__PURE__*/React.createElement(View, {
        style: styles.titleWrapper
      }, /*#__PURE__*/React.createElement(Text, {
        style: styles.titleText
      }, "Photos")), rightBtn)));
    }
  }]);

  return NavBar;
}(PureComponent);

NavBar.defaultProps = {
  styles: navBarStyles
  /* leftText: '返回',
  rightText: '确定',
  showRight: true, */

};
export { NavBar as default };
