// zarm/lib/cell/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _classnames3 = _interopRequireDefault(require("classnames"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var Cell = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(Cell, _PureComponent);

  var _super = _createSuper(Cell);

  function Cell() {
    (0, _classCallCheck2.default)(this, Cell);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Cell, [{
    key: "render",
    value: function render() {
      var _classnames;

      var _this$props = this.props,
          prefixCls = _this$props.prefixCls,
          className = _this$props.className,
          hasArrow = _this$props.hasArrow,
          icon = _this$props.icon,
          title = _this$props.title,
          description = _this$props.description,
          help = _this$props.help,
          disabled = _this$props.disabled,
          onClick = _this$props.onClick,
          children = _this$props.children,
          others = (0, _objectWithoutProperties2.default)(_this$props, ["prefixCls", "className", "hasArrow", "icon", "title", "description", "help", "disabled", "onClick", "children"]);
      var cls = (0, _classnames3.default)(prefixCls, className, (_classnames = {}, (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--disabled"), disabled), (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--link"), !disabled && !!onClick), (0, _defineProperty2.default)(_classnames, "".concat(prefixCls, "--arrow"), hasArrow), _classnames));
      var titleCls = (0, _classnames3.default)("".concat(prefixCls, "__title"), (0, _defineProperty2.default)({}, "".concat(prefixCls, "__title--label"), !!children));

      var iconRender = icon && /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__icon")
      }, icon);

      var titleRender = title && /*#__PURE__*/_react.default.createElement("div", {
        className: titleCls
      }, title);

      var contentRender = children && /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__content")
      }, children);

      var arrowRender = hasArrow && /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__arrow")
      });

      var helpRender = help && /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__help")
      }, help);

      return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
        className: cls,
        onClick: onClick,
        onTouchStart: function onTouchStart() {}
      }, others), /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__inner")
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__header")
      }, iconRender), /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__body")
      }, titleRender, contentRender), /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__footer")
      }, description), arrowRender), helpRender);
    }
  }]);
  return Cell;
}(_react.PureComponent);

exports.default = Cell;
Cell.defaultProps = {
  prefixCls: 'za-cell',
  hasArrow: false,
  disabled: false
};
