// zarm/lib/cell/index.native.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _index = _interopRequireDefault(require("./style/index.native"));

var _renderWithText = require("../utils/renderWithText.native");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var cellStyles = _reactNative.StyleSheet.create(_index.default);

var Cell = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(Cell, _PureComponent);

  var _super = _createSuper(Cell);

  function Cell(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Cell);
    _this = _super.call(this, props);

    _this.onPressIn = function () {
      _this.setState({
        isActive: true
      });
    };

    _this.onPressOut = function () {
      _this.setState({
        isActive: false
      });
    };

    _this.state = {
      isActive: false
    };
    return _this;
  }

  (0, _createClass2.default)(Cell, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          hasArrow = _this$props.hasArrow,
          icon = _this$props.icon,
          title = _this$props.title,
          description = _this$props.description,
          help = _this$props.help,
          style = _this$props.style,
          styles = _this$props.styles,
          onClick = _this$props.onClick,
          children = _this$props.children,
          others = (0, _objectWithoutProperties2.default)(_this$props, ["hasArrow", "icon", "title", "description", "help", "style", "styles", "onClick", "children"]);
      var isActive = this.state.isActive;
      var wrapperStyle = [styles.wrapperStyle, isActive && styles.activeWrapper, style];
      var titleViewStyle = [styles.titleViewStyle, !!children && styles.labelTitleViewStyle];
      var contentStyle = !onClick && wrapperStyle;

      var iconRender = icon && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: styles.iconStyle
      }, icon);

      var arrowRender = hasArrow && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: styles.arrowStyle
      });

      var cellRender = /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: wrapperStyle
      }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: styles.lineContainerStyle
      }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: styles.lineStyle
      })), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: styles.innerStyle
      }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: styles.contentStyle
      }, iconRender, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: styles.bodyStyle
      }, /*#__PURE__*/_react.default.createElement(_renderWithText.RenderWithText, {
        component: title,
        viewStyle: titleViewStyle,
        textStyle: styles.titleTextStyle
      }), /*#__PURE__*/_react.default.createElement(_renderWithText.RenderWithText, {
        component: children,
        viewStyle: styles.childrenViewStyle
      })), /*#__PURE__*/_react.default.createElement(_renderWithText.RenderWithText, {
        component: description,
        textStyle: styles.descriptionTextStyle
      }), arrowRender)), /*#__PURE__*/_react.default.createElement(_renderWithText.RenderWithText, {
        component: help,
        viewStyle: cellStyles.helpViewStyle
      }));

      var wrapperProps = _objectSpread({
        activeOpacity: 1,
        style: contentStyle,
        onPress: onClick,
        onPressIn: this.onPressIn,
        onPressOut: this.onPressOut
      }, others);

      return onClick ? /*#__PURE__*/_react.default.createElement(_reactNative.TouchableHighlight, wrapperProps, cellRender) : cellRender;
    }
  }]);
  return Cell;
}(_react.PureComponent);

exports.default = Cell;
Cell.defaultProps = {
  hasArrow: false,
  styles: cellStyles
};
