// zarm/lib/scroller/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = require("react");

var _events = _interopRequireDefault(require("../utils/events"));

var _throttle = _interopRequireDefault(require("../utils/throttle"));

var _dom = require("../utils/dom");

var _noop = _interopRequireDefault(require("../utils/noop"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var Scroller = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(Scroller, _PureComponent);

  var _super = _createSuper(Scroller);

  function Scroller() {
    var _this;

    (0, _classCallCheck2.default)(this, Scroller);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.mounted = void 0;
    _this.scrollThrottled = _this.props.onScroll ? (0, _throttle.default)(_this.props.onScroll, 250) : _noop.default;

    _this.onScroll = function () {
      var onScroll = _this.props.onScroll;
      if (!_this.mounted) return;
      typeof onScroll === 'function' && _this.scrollThrottled(_this.scrollTop);
    };

    _this.bindEvent = function () {
      _this.scrollContainer && _events.default.on(_this.scrollContainer, 'scroll', _this.onScroll);
    };

    _this.unBindEvent = function () {
      _this.scrollContainer && _events.default.off(_this.scrollContainer, 'scroll', _this.onScroll);
    };

    return _this;
  }

  (0, _createClass2.default)(Scroller, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.bindEvent();
      this.mounted = true;
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      var container = this.props.container;

      if (prevProps.container !== container) {
        this.bindEvent();
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.unBindEvent();
      this.mounted = false;
    }
  }, {
    key: "scrollContainer",
    get: function get() {
      var container = this.props.container;
      return (0, _dom.getScrollContainer)(container);
    }
  }, {
    key: "scrollTop",
    get: function get() {
      return (0, _dom.getScrollTop)(this.scrollContainer);
    }
  }, {
    key: "render",
    value: function render() {
      var children = this.props.children;
      return children || null;
    }
  }]);
  return Scroller;
}(_react.PureComponent);

exports.default = Scroller;
Scroller.displayName = 'Scroller';
Scroller.defaultProps = {
  prefixCls: 'za-scroller',
  container: _dom.canUseDOM ? window : undefined
};
