// zarm/es/calendar/utils/parseState.js
import DateTool from '../../utils/date';

var parseState = function parseState(props) {
  var defaultValue = props.defaultValue,
      multiple = props.multiple;
  var value = props.value;
  var tmpValue;
  value = value || defaultValue;
  value = Object.prototype.toString.call(value) === '[object Array]' ? value : value && [value] || []; // 注掉该逻辑，强制根据 multiple 控制节点个数，后面改进
  // tmpValue = value.map(item => DateTool.parseDay(item));

  tmpValue = value.slice(0, multiple ? 2 : 1).map(function (item) {
    return DateTool.parseDay(item);
  }); // 排序过滤

  tmpValue = tmpValue.sort(function (item1, item2) {
    return +item1 - +item2;
  });
  var min = props.min ? DateTool.parseDay(props.min) : new Date();
  var startMonth = DateTool.cloneDate(min, 'dd', 1);
  var max = props.max ? DateTool.parseDay(props.max) : DateTool.cloneDate(min, 'y', 1);
  var endMonth = DateTool.cloneDate(max, 'dd', DateTool.getDaysByDate(max)); // min、max 排序

  var duration = [min, max].sort(function (item1, item2) {
    return +item1 - +item2;
  });
  var tmp = {
    value: tmpValue,
    min: duration[0],
    max: duration[1],
    startMonth: startMonth,
    endMonth: endMonth,
    // 是否是入参更新(主要是月份跨度更新，需要重新定位)
    refresh: false,
    // 注掉该逻辑，强制根据 multiple 控制节点个数，后面改进
    // steps:Math.max(tmp.value.length, tmp.defaultValue.length);
    steps: multiple ? 2 : 1,
    // 初始化点击步数
    multiple: multiple
  };
  return tmp;
};

export default parseState;
