// zarm/lib/keyboard/Keyboard.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _icons = require("@zarm-design/icons");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var KEYS = {
  number: ['1', '2', '3', '4', '5', '6', '7', '8', '9', '', '0', 'close'],
  price: ['1', '2', '3', '4', '5', '6', '7', '8', '9', '.', '0', 'close'],
  idcard: ['1', '2', '3', '4', '5', '6', '7', '8', '9', 'X', '0', 'close']
};

var Keyboard = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(Keyboard, _PureComponent);

  var _super = _createSuper(Keyboard);

  function Keyboard() {
    var _this;

    (0, _classCallCheck2.default)(this, Keyboard);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.touchLongPressTimer = void 0;
    _this.mouseLongPressTimer = void 0;

    _this.onTouchLongPressIn = function (key) {
      _this.onKeyClick(key);

      _this.touchLongPressTimer = setTimeout(function () {
        _this.touchLongPressTimer = setInterval(function () {
          _this.onKeyClick(key);
        }, 100);
      }, 800);
    };

    _this.onTouchLongPressOut = function (e) {
      e.preventDefault();
      clearInterval(_this.touchLongPressTimer);
    };

    _this.onMouseLongPressIn = function (key) {
      _this.onKeyClick(key);

      _this.mouseLongPressTimer = setTimeout(function () {
        _this.mouseLongPressTimer = setInterval(function () {
          _this.onKeyClick(key);
        }, 100);
      }, 800);
    };

    _this.onMouseLongPressOut = function (e) {
      e.preventDefault();
      clearInterval(_this.mouseLongPressTimer);
    };

    _this.onKeyClick = function (key) {
      if (key.length === 0) {
        return;
      }

      var onKeyClick = _this.props.onKeyClick;

      if (typeof onKeyClick === 'function') {
        onKeyClick(key);
      }
    };

    _this.getKeys = function () {
      var type = _this.props.type;
      return type ? KEYS[type] : KEYS.number;
    };

    _this.renderKey = function (text, index) {
      var prefixCls = _this.props.prefixCls;
      var keyCls = (0, _classnames2.default)("".concat(prefixCls, "__item"), (0, _defineProperty2.default)({}, "".concat(prefixCls, "__item--disabled"), text.length === 0));
      return /*#__PURE__*/_react.default.createElement("div", {
        className: keyCls,
        key: +index,
        onClick: function onClick() {
          return _this.onKeyClick(text);
        }
      }, text === 'close' ? /*#__PURE__*/_react.default.createElement(_icons.Keyboard, {
        size: "lg"
      }) : text);
    };

    return _this;
  }

  (0, _createClass2.default)(Keyboard, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          prefixCls = _this$props.prefixCls,
          className = _this$props.className,
          locale = _this$props.locale;
      var cls = (0, _classnames2.default)(prefixCls, className);
      return /*#__PURE__*/_react.default.createElement("div", {
        className: cls
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__keys")
      }, this.getKeys().map(this.renderKey)), /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__handle")
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__item"),
        onTouchStart: function onTouchStart() {
          return _this2.onTouchLongPressIn('delete');
        },
        onTouchEnd: this.onTouchLongPressOut,
        onTouchCancel: this.onTouchLongPressOut,
        onMouseDown: function onMouseDown() {
          return _this2.onMouseLongPressIn('delete');
        },
        onMouseUp: this.onMouseLongPressOut
      }, /*#__PURE__*/_react.default.createElement(_icons.DeleteKey, {
        size: "lg"
      })), /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "__item ").concat(prefixCls, "__item--ok"),
        onClick: function onClick() {
          return _this2.onKeyClick('ok');
        }
      }, locale.okText)));
    }
  }]);
  return Keyboard;
}(_react.PureComponent);

exports.default = Keyboard;
Keyboard.defaultProps = {
  prefixCls: 'za-keyboard',
  type: 'number'
};
