// zarm/es/action-sheet/index.native.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import React, { PureComponent } from 'react';
import { StyleSheet, Text, View, TouchableOpacity } from 'react-native';
import actionsheetStyle from './style/index.native';
import Popup from '../popup/index.native';
var actionsheetStyles = StyleSheet.create(actionsheetStyle);

var ActionSheet = /*#__PURE__*/function (_PureComponent) {
  _inherits(ActionSheet, _PureComponent);

  var _super = _createSuper(ActionSheet);

  function ActionSheet() {
    var _this;

    _classCallCheck(this, ActionSheet);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.state = {
      isActive: false,
      activeIndex: null
    };

    _this.onPressIn = function (activeIndex) {
      _this.setState({
        isActive: true,
        activeIndex: activeIndex
      });
    };

    _this.onPressOut = function (activeIndex) {
      _this.setState({
        isActive: false,
        activeIndex: activeIndex
      });
    };

    _this.renderActions = function (action, index) {
      var styles = _this.props.styles;
      var _this$state = _this.state,
          isActive = _this$state.isActive,
          activeIndex = _this$state.activeIndex;
      var actionMaskStyle = [// index !== 0 && styles!.actionItemBorder,
      styles.actionItemMask, activeIndex === index && isActive && styles.actionItemActive];
      var actionStyle = [styles.actionItem, !!action.theme && styles["theme-".concat(action.theme)]];
      return /*#__PURE__*/React.createElement(TouchableOpacity, {
        key: +index,
        activeOpacity: 1,
        onPress: action.onClick,
        onPressIn: function onPressIn() {
          return _this.onPressIn(index);
        },
        onPressOut: function onPressOut() {
          return _this.onPressOut(index);
        }
      }, /*#__PURE__*/React.createElement(View, {
        style: actionMaskStyle
      }), index !== 0 ? /*#__PURE__*/React.createElement(View, {
        style: styles.actionItemBorder
      }) : null, /*#__PURE__*/React.createElement(Text, {
        style: actionStyle
      }, action.text));
    };

    _this.renderCancel = function () {
      var _this$props = _this.props,
          styles = _this$props.styles,
          onCancel = _this$props.onCancel,
          cancelText = _this$props.cancelText;
      var _this$state2 = _this.state,
          isActive = _this$state2.isActive,
          activeIndex = _this$state2.activeIndex;
      var cancelWrapperStyle = [styles.wrapperCacnel];
      var cancelMaskStyle = [styles.actionItemMask, isActive && activeIndex === 'cancel' && styles.actionItemActive];
      var cancelStyle = [styles.actionCancelItem];
      return typeof onCancel === 'function' && /*#__PURE__*/React.createElement(TouchableOpacity, {
        style: cancelWrapperStyle,
        activeOpacity: 1,
        onPress: onCancel,
        onPressIn: function onPressIn() {
          return _this.onPressIn('cancel');
        },
        onPressOut: function onPressOut() {
          return _this.onPressOut('cancel');
        }
      }, /*#__PURE__*/React.createElement(View, {
        style: cancelMaskStyle
      }), /*#__PURE__*/React.createElement(Text, {
        style: cancelStyle
      }, cancelText));
    };

    return _this;
  }

  _createClass(ActionSheet, [{
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          style = _this$props2.style,
          styles = _this$props2.styles,
          spacing = _this$props2.spacing,
          visible = _this$props2.visible,
          onMaskClick = _this$props2.onMaskClick,
          actions = _this$props2.actions;
      var wrapperStyle = [styles.wrapper, spacing && styles.wrapperSpacing, style];
      var actionsStyle = [styles.wrapperActions, {
        margin: 0
      }];
      return /*#__PURE__*/React.createElement(Popup, {
        visible: visible,
        onMaskClick: onMaskClick
      }, /*#__PURE__*/React.createElement(View, {
        style: wrapperStyle
      }, /*#__PURE__*/React.createElement(View, {
        style: actionsStyle
      }, actions.map(this.renderActions)), this.renderCancel()));
    }
  }]);

  return ActionSheet;
}(PureComponent);

ActionSheet.defaultProps = {
  visible: false,
  spacing: false,
  actions: [],
  cancelText: '取消',
  styles: actionsheetStyles
};
export { ActionSheet as default };
