// zarm/lib/calendar/utils/parseState.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _date = _interopRequireDefault(require("../../utils/date"));

var parseState = function parseState(props) {
  var defaultValue = props.defaultValue,
      multiple = props.multiple;
  var value = props.value;
  var tmpValue;
  value = value || defaultValue;
  value = Object.prototype.toString.call(value) === '[object Array]' ? value : value && [value] || []; // 注掉该逻辑，强制根据 multiple 控制节点个数，后面改进
  // tmpValue = value.map(item => DateTool.parseDay(item));

  tmpValue = value.slice(0, multiple ? 2 : 1).map(function (item) {
    return _date.default.parseDay(item);
  }); // 排序过滤

  tmpValue = tmpValue.sort(function (item1, item2) {
    return +item1 - +item2;
  });
  var min = props.min ? _date.default.parseDay(props.min) : new Date();

  var startMonth = _date.default.cloneDate(min, 'dd', 1);

  var max = props.max ? _date.default.parseDay(props.max) : _date.default.cloneDate(min, 'y', 1);

  var endMonth = _date.default.cloneDate(max, 'dd', _date.default.getDaysByDate(max)); // min、max 排序


  var duration = [min, max].sort(function (item1, item2) {
    return +item1 - +item2;
  });
  var tmp = {
    value: tmpValue,
    min: duration[0],
    max: duration[1],
    startMonth: startMonth,
    endMonth: endMonth,
    // 是否是入参更新(主要是月份跨度更新，需要重新定位)
    refresh: false,
    // 注掉该逻辑，强制根据 multiple 控制节点个数，后面改进
    // steps:Math.max(tmp.value.length, tmp.defaultValue.length);
    steps: multiple ? 2 : 1,
    // 初始化点击步数
    multiple: multiple
  };
  return tmp;
};

var _default = parseState;
exports.default = _default;
